{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Visualisation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook aims to explore the data retrived from ENCODE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  File accession   File format File type File format type  \\\n",
      "0    ENCFF690FNR     bed bed9+       bed            bed9+   \n",
      "1    ENCFF884UAE        bigWig    bigWig              NaN   \n",
      "2    ENCFF464CWC        bigWig    bigWig              NaN   \n",
      "3    ENCFF549SNG  bigBed bed9+    bigBed            bed9+   \n",
      "4    ENCFF840XVU     bed bed9+       bed            bed9+   \n",
      "\n",
      "                             Output type File assembly Experiment accession  \\\n",
      "0               methylation state at CpG        GRCh38          ENCSR786DCL   \n",
      "1   plus strand methylation state at CpG        GRCh38          ENCSR786DCL   \n",
      "2  minus strand methylation state at CpG        GRCh38          ENCSR786DCL   \n",
      "3               methylation state at CpG        GRCh38          ENCSR786DCL   \n",
      "4               methylation state at CpG        GRCh38          ENCSR625HZA   \n",
      "\n",
      "  Assay                    Donor(s) Biosample term id  ... Platform  \\\n",
      "0  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   \n",
      "1  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   \n",
      "2  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   \n",
      "3  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   \n",
      "4  WGBS  /human-donors/ENCDO336AAA/       EFO:0005723  ...      NaN   \n",
      "\n",
      "  Controlled by File Status  \\\n",
      "0           NaN    released   \n",
      "1           NaN    released   \n",
      "2           NaN    released   \n",
      "3           NaN    released   \n",
      "4           NaN    released   \n",
      "\n",
      "                                              s3_uri  \\\n",
      "0  s3://encode-public/2021/07/17/dccc66ef-985f-4e...   \n",
      "1  s3://encode-public/2021/07/17/8407f321-a9f9-4d...   \n",
      "2  s3://encode-public/2021/07/17/b72fba8b-4ccf-47...   \n",
      "3  s3://encode-public/2021/07/17/6aff99be-3685-4a...   \n",
      "4  s3://encode-public/2021/07/20/bfc10eed-ce81-48...   \n",
      "\n",
      "                                           Azure URL    File analysis title  \\\n",
      "0  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "1  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "2  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "3  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "4  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "\n",
      "   File analysis status                 Audit WARNING  Audit NOT_COMPLIANT  \\\n",
      "0              released  low lambda C conversion rate                  NaN   \n",
      "1              released  low lambda C conversion rate                  NaN   \n",
      "2              released  low lambda C conversion rate                  NaN   \n",
      "3              released  low lambda C conversion rate                  NaN   \n",
      "4              released  low lambda C conversion rate                  NaN   \n",
      "\n",
      "                        Audit ERROR  \n",
      "0                               NaN  \n",
      "1                               NaN  \n",
      "2                               NaN  \n",
      "3                               NaN  \n",
      "4  missing lambda C conversion rate  \n",
      "\n",
      "[5 rows x 59 columns]\n"
     ]
    }
   ],
   "source": [
    "# loading the data\n",
    "import pandas as pd\n",
    "\n",
    "metadata_df = pd.read_csv('HepG2_data/HepG2_DNAm/metadata.tsv', sep='\\t')\n",
    "print(metadata_df.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "WGBS - methylation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     0      1      2   3   4  5      6      7        8   9    10  11  12  13\n",
      "0  chr1  10468  10469   1   1  +  10468  10469  0,255,0   1    0  CG  CN   1\n",
      "1  chr1  10496  10497   1   1  +  10496  10497  255,0,0   1  100  CG  CG   2\n",
      "2  chr1  10524  10525   1   1  +  10524  10525  255,0,0   1  100  CG  CG   2\n",
      "3  chr1  10541  10542   1   1  +  10541  10542  255,0,0   1  100  CG  CG   2\n",
      "4  chr1  10562  10563   1   1  +  10562  10563  255,0,0   1  100  CG  CG   2\n",
      "5  chr1  10570  10571   1   1  +  10570  10571  255,0,0   1  100  CG  CG   2\n",
      "6  chr1  10576  10577   1   1  +  10576  10577  255,0,0   1  100  CG  CG   2\n",
      "7  chr1  10578  10579   1   1  +  10578  10579  255,0,0   1  100  CG  CG   2\n",
      "8  chr1  10588  10589   1   1  +  10588  10589  255,0,0   1  100  CG  CG   2\n",
      "9  chr1  10608  10609   1   1  +  10608  10609  255,0,0   1  100  CG  CG   2\n"
     ]
    }
   ],
   "source": [
    "WGBS_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF857YRG.bed', sep='\\t', nrows=10, header =None)\n",
    "print(WGBS_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "Unable to allocate 512. KiB for an array with shape (65536,) and data type int64",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[19], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# loading WGBS data (methylation)\u001b[39;00m\n\u001b[1;32m----> 3\u001b[0m WGBS_df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mHepG2_data/HepG2_DNAm/processed/ENCFF690FNR.bed\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnames\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrefChr\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mstartChr\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mendChr\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mname\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mscore\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mstrand\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mstartThick\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mendThick\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrgb\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mreads\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpctMe\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrefGen\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msampGen\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mqualScore\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m      4\u001b[0m WGBS_plus_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHepG2_data/HepG2_DNAm/processed/ENCFF884UAE.bedGraph\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, names\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrefChr\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstart\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mend\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpctMe\u001b[39m\u001b[38;5;124m'\u001b[39m], header\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m      5\u001b[0m WGBS_minus_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHepG2_data/HepG2_DNAm/processed/ENCFF464CWC.bedGraph\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, names\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrefChr\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstart\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mend\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpctMe\u001b[39m\u001b[38;5;124m'\u001b[39m], header\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m)\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\util\\_decorators.py:211\u001b[0m, in \u001b[0;36mdeprecate_kwarg.<locals>._deprecate_kwarg.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    209\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    210\u001b[0m         kwargs[new_arg_name] \u001b[38;5;241m=\u001b[39m new_arg_value\n\u001b[1;32m--> 211\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\util\\_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    325\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[0;32m    326\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[0;32m    327\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[0;32m    328\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[0;32m    329\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[0;32m    330\u001b[0m     )\n\u001b[1;32m--> 331\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:950\u001b[0m, in \u001b[0;36mread_csv\u001b[1;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[0;32m    935\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[0;32m    936\u001b[0m     dialect,\n\u001b[0;32m    937\u001b[0m     delimiter,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    946\u001b[0m     defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[0;32m    947\u001b[0m )\n\u001b[0;32m    948\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[1;32m--> 950\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:611\u001b[0m, in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    608\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n\u001b[0;32m    610\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[1;32m--> 611\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:1778\u001b[0m, in \u001b[0;36mTextFileReader.read\u001b[1;34m(self, nrows)\u001b[0m\n\u001b[0;32m   1771\u001b[0m nrows \u001b[38;5;241m=\u001b[39m validate_integer(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnrows\u001b[39m\u001b[38;5;124m\"\u001b[39m, nrows)\n\u001b[0;32m   1772\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1773\u001b[0m     \u001b[38;5;66;03m# error: \"ParserBase\" has no attribute \"read\"\u001b[39;00m\n\u001b[0;32m   1774\u001b[0m     (\n\u001b[0;32m   1775\u001b[0m         index,\n\u001b[0;32m   1776\u001b[0m         columns,\n\u001b[0;32m   1777\u001b[0m         col_dict,\n\u001b[1;32m-> 1778\u001b[0m     ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[attr-defined]\u001b[39;49;00m\n\u001b[0;32m   1779\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\n\u001b[0;32m   1780\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1781\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[0;32m   1782\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\c_parser_wrapper.py:230\u001b[0m, in \u001b[0;36mCParserWrapper.read\u001b[1;34m(self, nrows)\u001b[0m\n\u001b[0;32m    228\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    229\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlow_memory:\n\u001b[1;32m--> 230\u001b[0m         chunks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader\u001b[38;5;241m.\u001b[39mread_low_memory(nrows)\n\u001b[0;32m    231\u001b[0m         \u001b[38;5;66;03m# destructive to chunks\u001b[39;00m\n\u001b[0;32m    232\u001b[0m         data \u001b[38;5;241m=\u001b[39m _concatenate_chunks(chunks)\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:808\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.read_low_memory\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:890\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._read_rows\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:1037\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._convert_column_data\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:1083\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._convert_tokens\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:1183\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._convert_with_dtype\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:1744\u001b[0m, in \u001b[0;36mpandas._libs.parsers._try_int64\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mMemoryError\u001b[0m: Unable to allocate 512. KiB for an array with shape (65536,) and data type int64"
     ]
    }
   ],
   "source": [
    "# loading WGBS data (methylation)\n",
    "\n",
    "WGBS_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF690FNR.bed', sep='\\t', names=['refChr', 'startChr', 'endChr', 'name', 'score', 'strand', 'startThick', 'endThick', 'rgb', 'reads', 'pctMe', 'refGen', 'sampGen', 'qualScore'], header =None)\n",
    "WGBS_plus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF884UAE.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'pctMe'], header=None)\n",
    "WGBS_minus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF464CWC.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'pctMe'], header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns in big bed files (in FNR bed file):\n",
    "- Reference chromosome or scaffold\n",
    "- Start position in chromosome\n",
    "- End position in chromosome\n",
    "- Name of item\n",
    "- Score from 0-1000. Capped number of reads\n",
    "- Strandedness, plus (+), minus (-), or unknown (?)\n",
    "- Start of where display should be thick\n",
    "- End of where display should be thick\n",
    "- Color value (RGB)\n",
    "- Coverage, or number of reads\n",
    "- Percentage of reads that show methylation at this position in the genome\n",
    "- Reference genotype\n",
    "- Sample genotype\n",
    "- Quality score for genotype call\n",
    "\n",
    "\n",
    "Final column in bigwig files = percent methylation at CpG site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>startChr</th>\n",
       "      <th>endChr</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>startThick</th>\n",
       "      <th>endThick</th>\n",
       "      <th>rgb</th>\n",
       "      <th>reads</th>\n",
       "      <th>pctMe</th>\n",
       "      <th>refGen</th>\n",
       "      <th>sampGen</th>\n",
       "      <th>qualScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CG</td>\n",
       "      <td>MG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>155,255,0</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>CG</td>\n",
       "      <td>CK</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refChr  startChr  endChr  name  score strand  startThick  endThick  \\\n",
       "0   chr1     10468   10469     1      1      +       10468     10469   \n",
       "1   chr1     10469   10470     1      4      -       10469     10470   \n",
       "2   chr1     10470   10471     1      1      +       10470     10471   \n",
       "3   chr1     10471   10472     1      2      -       10471     10472   \n",
       "4   chr1     10483   10484     1      1      +       10483     10484   \n",
       "\n",
       "         rgb  reads  pctMe refGen sampGen  qualScore  \n",
       "0    0,255,0      1      0     CG      MG          3  \n",
       "1  155,255,0      4     25     CG      CK          5  \n",
       "2    255,0,0      1    100     CG      CG         11  \n",
       "3    255,0,0      2    100     CG      CG          6  \n",
       "4    255,0,0      1    100     CG      CG          4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>pctMe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10009</td>\n",
       "      <td>10012</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10015</td>\n",
       "      <td>10018</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10021</td>\n",
       "      <td>10024</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10027</td>\n",
       "      <td>10030</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10033</td>\n",
       "      <td>10036</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refChr  start    end  pctMe\n",
       "0   chr1  10009  10012    0.0\n",
       "1   chr1  10015  10018    0.0\n",
       "2   chr1  10021  10024    0.0\n",
       "3   chr1  10027  10030    0.0\n",
       "4   chr1  10033  10036    0.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>pctMe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10472</td>\n",
       "      <td>10473</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10481</td>\n",
       "      <td>10482</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10484</td>\n",
       "      <td>10485</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refChr  start    end  pctMe\n",
       "0   chr1  10469  10470   25.0\n",
       "1   chr1  10471  10472  100.0\n",
       "2   chr1  10472  10473    0.0\n",
       "3   chr1  10481  10482    0.0\n",
       "4   chr1  10484  10485  100.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(WGBS_df.head())\n",
    "display(WGBS_plus_df.head())\n",
    "display(WGBS_minus_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA data - plus/minus strand and gene quant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#loading RNA data\n",
    "\n",
    "RNA_geneQuant_df = pd.read_csv('HepG2_data/HepG2_DNAm/ENCFF649XOG.tsv', sep='\\t', header=0)\n",
    "RNA_minus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF194TSC.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'exp'])\n",
    "RNA_plus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF792BNC.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'exp'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns for RNA:\n",
    "\n",
    "- chromosome\n",
    "- start\n",
    "- end\n",
    "- gene expression?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190711</td>\n",
       "      <td>190761</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190786</td>\n",
       "      <td>190808</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190808</td>\n",
       "      <td>190836</td>\n",
       "      <td>0.06349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190836</td>\n",
       "      <td>190858</td>\n",
       "      <td>0.04233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190864</td>\n",
       "      <td>190892</td>\n",
       "      <td>0.04233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198639</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22823877</td>\n",
       "      <td>22823927</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198640</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22824904</td>\n",
       "      <td>22824954</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198641</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22825018</td>\n",
       "      <td>22825067</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198642</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56871389</td>\n",
       "      <td>56871439</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198643</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56871467</td>\n",
       "      <td>56871517</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7198644 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        refChr     start       end      exp\n",
       "0         chr1    190711    190761  0.02116\n",
       "1         chr1    190786    190808  0.02116\n",
       "2         chr1    190808    190836  0.06349\n",
       "3         chr1    190836    190858  0.04233\n",
       "4         chr1    190864    190892  0.04233\n",
       "...        ...       ...       ...      ...\n",
       "7198639   chrY  22823877  22823927  0.02116\n",
       "7198640   chrY  22824904  22824954  0.02116\n",
       "7198641   chrY  22825018  22825067  0.02116\n",
       "7198642   chrY  56871389  56871439  0.02116\n",
       "7198643   chrY  56871467  56871517  0.02116\n",
       "\n",
       "[7198644 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14475</td>\n",
       "      <td>14490</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14490</td>\n",
       "      <td>14525</td>\n",
       "      <td>0.04233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14525</td>\n",
       "      <td>14540</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14660</td>\n",
       "      <td>14668</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14668</td>\n",
       "      <td>14675</td>\n",
       "      <td>0.04233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063920</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24213384</td>\n",
       "      <td>24213405</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063921</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24213792</td>\n",
       "      <td>24213821</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063922</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26621781</td>\n",
       "      <td>26621831</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063923</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26621878</td>\n",
       "      <td>26621895</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063924</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26622511</td>\n",
       "      <td>26622544</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7063925 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        refChr     start       end      exp\n",
       "0         chr1     14475     14490  0.02116\n",
       "1         chr1     14490     14525  0.04233\n",
       "2         chr1     14525     14540  0.02116\n",
       "3         chr1     14660     14668  0.02116\n",
       "4         chr1     14668     14675  0.04233\n",
       "...        ...       ...       ...      ...\n",
       "7063920   chrY  24213384  24213405  0.02116\n",
       "7063921   chrY  24213792  24213821  0.02116\n",
       "7063922   chrY  26621781  26621831  0.02116\n",
       "7063923   chrY  26621878  26621895  0.02116\n",
       "7063924   chrY  26622511  26622544  0.02116\n",
       "\n",
       "[7063925 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10904</td>\n",
       "      <td>10904</td>\n",
       "      <td>93.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12954</td>\n",
       "      <td>12954</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12956</td>\n",
       "      <td>12956</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12958</td>\n",
       "      <td>12958</td>\n",
       "      <td>82.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12960</td>\n",
       "      <td>12960</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59521</th>\n",
       "      <td>gSpikein_ERCC-00165</td>\n",
       "      <td>tSpikein_ERCC-00165</td>\n",
       "      <td>872.0</td>\n",
       "      <td>700.78</td>\n",
       "      <td>87.0</td>\n",
       "      <td>4.47</td>\n",
       "      <td>8.43</td>\n",
       "      <td>87.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.33</td>\n",
       "      <td>8.42</td>\n",
       "      <td>3.439680</td>\n",
       "      <td>5.244820</td>\n",
       "      <td>0.071704</td>\n",
       "      <td>6.672670</td>\n",
       "      <td>10.182700</td>\n",
       "      <td>0.071661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59522</th>\n",
       "      <td>gSpikein_ERCC-00168</td>\n",
       "      <td>tSpikein_ERCC-00168</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>852.78</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.08</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.002566</td>\n",
       "      <td>0.194005</td>\n",
       "      <td>0.474234</td>\n",
       "      <td>0.005171</td>\n",
       "      <td>0.377749</td>\n",
       "      <td>0.474374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59523</th>\n",
       "      <td>gSpikein_ERCC-00170</td>\n",
       "      <td>tSpikein_ERCC-00170</td>\n",
       "      <td>1023.0</td>\n",
       "      <td>851.78</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.16</td>\n",
       "      <td>4.07</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.10</td>\n",
       "      <td>4.09</td>\n",
       "      <td>1.548640</td>\n",
       "      <td>2.687830</td>\n",
       "      <td>0.093629</td>\n",
       "      <td>3.011470</td>\n",
       "      <td>5.228390</td>\n",
       "      <td>0.093579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59524</th>\n",
       "      <td>gSpikein_ERCC-00171</td>\n",
       "      <td>tSpikein_ERCC-00171</td>\n",
       "      <td>505.0</td>\n",
       "      <td>333.85</td>\n",
       "      <td>3942.0</td>\n",
       "      <td>425.45</td>\n",
       "      <td>802.16</td>\n",
       "      <td>3942.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>406.87</td>\n",
       "      <td>791.53</td>\n",
       "      <td>394.250000</td>\n",
       "      <td>419.834000</td>\n",
       "      <td>0.010739</td>\n",
       "      <td>767.089000</td>\n",
       "      <td>816.791000</td>\n",
       "      <td>0.010720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59525</th>\n",
       "      <td>gSpikein_phiX174</td>\n",
       "      <td>tSpikein_phiX174</td>\n",
       "      <td>5386.0</td>\n",
       "      <td>5214.78</td>\n",
       "      <td>78337.0</td>\n",
       "      <td>541.27</td>\n",
       "      <td>1020.55</td>\n",
       "      <td>78337.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>517.51</td>\n",
       "      <td>1006.78</td>\n",
       "      <td>513.704000</td>\n",
       "      <td>521.421000</td>\n",
       "      <td>0.002592</td>\n",
       "      <td>999.854000</td>\n",
       "      <td>1013.850000</td>\n",
       "      <td>0.002408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59526 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id     transcript_id(s)  length  effective_length  \\\n",
       "0                    10904                10904    93.0              0.00   \n",
       "1                    12954                12954    94.0              0.00   \n",
       "2                    12956                12956    72.0              0.00   \n",
       "3                    12958                12958    82.0              0.00   \n",
       "4                    12960                12960    73.0              0.00   \n",
       "...                    ...                  ...     ...               ...   \n",
       "59521  gSpikein_ERCC-00165  tSpikein_ERCC-00165   872.0            700.78   \n",
       "59522  gSpikein_ERCC-00168  tSpikein_ERCC-00168  1024.0            852.78   \n",
       "59523  gSpikein_ERCC-00170  tSpikein_ERCC-00170  1023.0            851.78   \n",
       "59524  gSpikein_ERCC-00171  tSpikein_ERCC-00171   505.0            333.85   \n",
       "59525     gSpikein_phiX174     tSpikein_phiX174  5386.0           5214.78   \n",
       "\n",
       "       expected_count     TPM     FPKM  posterior_mean_count  \\\n",
       "0                 0.0    0.00     0.00                   0.0   \n",
       "1                 0.0    0.00     0.00                   0.0   \n",
       "2                 0.0    0.00     0.00                   0.0   \n",
       "3                 0.0    0.00     0.00                   0.0   \n",
       "4                 0.0    0.00     0.00                   0.0   \n",
       "...               ...     ...      ...                   ...   \n",
       "59521            87.0    4.47     8.43                  87.0   \n",
       "59522             1.0    0.04     0.08                   1.0   \n",
       "59523            51.0    2.16     4.07                  51.0   \n",
       "59524          3942.0  425.45   802.16                3942.0   \n",
       "59525         78337.0  541.27  1020.55               78337.0   \n",
       "\n",
       "       posterior_standard_deviation_of_count  pme_TPM  pme_FPKM  \\\n",
       "0                                        0.0     0.00      0.00   \n",
       "1                                        0.0     0.00      0.00   \n",
       "2                                        0.0     0.00      0.00   \n",
       "3                                        0.0     0.00      0.00   \n",
       "4                                        0.0     0.00      0.00   \n",
       "...                                      ...      ...       ...   \n",
       "59521                                    0.0     4.33      8.42   \n",
       "59522                                    0.0     0.08      0.16   \n",
       "59523                                    0.0     2.10      4.09   \n",
       "59524                                    0.0   406.87    791.53   \n",
       "59525                                    0.0   517.51   1006.78   \n",
       "\n",
       "       TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0                0.000000            0.000000   \n",
       "1                0.000000            0.000000   \n",
       "2                0.000000            0.000000   \n",
       "3                0.000000            0.000000   \n",
       "4                0.000000            0.000000   \n",
       "...                   ...                 ...   \n",
       "59521            3.439680            5.244820   \n",
       "59522            0.002566            0.194005   \n",
       "59523            1.548640            2.687830   \n",
       "59524          394.250000          419.834000   \n",
       "59525          513.704000          521.421000   \n",
       "\n",
       "       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                   0.000000             0.000000   \n",
       "1                                   0.000000             0.000000   \n",
       "2                                   0.000000             0.000000   \n",
       "3                                   0.000000             0.000000   \n",
       "4                                   0.000000             0.000000   \n",
       "...                                      ...                  ...   \n",
       "59521                               0.071704             6.672670   \n",
       "59522                               0.474234             0.005171   \n",
       "59523                               0.093629             3.011470   \n",
       "59524                               0.010739           767.089000   \n",
       "59525                               0.002592           999.854000   \n",
       "\n",
       "       FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \n",
       "0                 0.000000                                0.000000  \n",
       "1                 0.000000                                0.000000  \n",
       "2                 0.000000                                0.000000  \n",
       "3                 0.000000                                0.000000  \n",
       "4                 0.000000                                0.000000  \n",
       "...                    ...                                     ...  \n",
       "59521            10.182700                                0.071661  \n",
       "59522             0.377749                                0.474374  \n",
       "59523             5.228390                                0.093579  \n",
       "59524           816.791000                                0.010720  \n",
       "59525          1013.850000                                0.002408  \n",
       "\n",
       "[59526 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(RNA_plus_df)\n",
    "display(RNA_minus_df)\n",
    "display(RNA_geneQuant_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10000</th>\n",
       "      <td>ENSG00000151006.7</td>\n",
       "      <td>ENST00000280606.6,ENST00000486499.1,ENST000004...</td>\n",
       "      <td>4510.90</td>\n",
       "      <td>4339.68</td>\n",
       "      <td>124.04</td>\n",
       "      <td>1.03</td>\n",
       "      <td>1.94</td>\n",
       "      <td>134.12</td>\n",
       "      <td>6.64</td>\n",
       "      <td>1.21</td>\n",
       "      <td>2.36</td>\n",
       "      <td>0.872066</td>\n",
       "      <td>1.566580</td>\n",
       "      <td>0.102375</td>\n",
       "      <td>1.704870</td>\n",
       "      <td>3.056580</td>\n",
       "      <td>0.102409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10001</th>\n",
       "      <td>ENSG00000151012.13</td>\n",
       "      <td>ENST00000280612.9,ENST00000509248.1</td>\n",
       "      <td>9249.21</td>\n",
       "      <td>9077.99</td>\n",
       "      <td>1848.00</td>\n",
       "      <td>7.33</td>\n",
       "      <td>13.83</td>\n",
       "      <td>1848.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7.14</td>\n",
       "      <td>13.88</td>\n",
       "      <td>6.625750</td>\n",
       "      <td>7.708410</td>\n",
       "      <td>0.025499</td>\n",
       "      <td>12.886400</td>\n",
       "      <td>14.991900</td>\n",
       "      <td>0.025553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10002</th>\n",
       "      <td>ENSG00000151014.5</td>\n",
       "      <td>ENST00000280614.3,ENST00000515616.1,ENST000006...</td>\n",
       "      <td>1856.43</td>\n",
       "      <td>1685.21</td>\n",
       "      <td>115.00</td>\n",
       "      <td>2.46</td>\n",
       "      <td>4.64</td>\n",
       "      <td>115.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.63</td>\n",
       "      <td>5.11</td>\n",
       "      <td>2.013440</td>\n",
       "      <td>3.298030</td>\n",
       "      <td>0.083487</td>\n",
       "      <td>3.915400</td>\n",
       "      <td>6.415060</td>\n",
       "      <td>0.083588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10003</th>\n",
       "      <td>ENSG00000151023.16</td>\n",
       "      <td>ENST00000331161.8,ENST00000376363.5,ENST000004...</td>\n",
       "      <td>1318.00</td>\n",
       "      <td>1146.78</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.12</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.048159</td>\n",
       "      <td>0.425282</td>\n",
       "      <td>0.317992</td>\n",
       "      <td>0.093792</td>\n",
       "      <td>0.827459</td>\n",
       "      <td>0.317884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10004</th>\n",
       "      <td>ENSG00000151025.10</td>\n",
       "      <td>ENST00000376351.3,ENST00000482641.2,ENST000004...</td>\n",
       "      <td>4853.25</td>\n",
       "      <td>4682.04</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.004778</td>\n",
       "      <td>0.197461</td>\n",
       "      <td>0.489654</td>\n",
       "      <td>0.010009</td>\n",
       "      <td>0.384865</td>\n",
       "      <td>0.489683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10995</th>\n",
       "      <td>ENSG00000159387.7</td>\n",
       "      <td>ENST00000290552.7,ENST00000558315.1</td>\n",
       "      <td>1989.00</td>\n",
       "      <td>1817.78</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.000902</td>\n",
       "      <td>0.167404</td>\n",
       "      <td>0.532145</td>\n",
       "      <td>0.001434</td>\n",
       "      <td>0.325286</td>\n",
       "      <td>0.532105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10996</th>\n",
       "      <td>ENSG00000159388.5</td>\n",
       "      <td>ENST00000290551.4,ENST00000475157.1</td>\n",
       "      <td>2579.06</td>\n",
       "      <td>2407.84</td>\n",
       "      <td>286.00</td>\n",
       "      <td>4.28</td>\n",
       "      <td>8.07</td>\n",
       "      <td>286.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.15</td>\n",
       "      <td>8.07</td>\n",
       "      <td>3.643900</td>\n",
       "      <td>4.658430</td>\n",
       "      <td>0.042490</td>\n",
       "      <td>7.082570</td>\n",
       "      <td>9.055910</td>\n",
       "      <td>0.042501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10997</th>\n",
       "      <td>ENSG00000159398.15</td>\n",
       "      <td>ENST00000290567.13,ENST00000319165.13,ENST0000...</td>\n",
       "      <td>1766.22</td>\n",
       "      <td>1595.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.085282</td>\n",
       "      <td>0.469493</td>\n",
       "      <td>0.262764</td>\n",
       "      <td>0.157796</td>\n",
       "      <td>0.905380</td>\n",
       "      <td>0.262659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10998</th>\n",
       "      <td>ENSG00000159399.9</td>\n",
       "      <td>ENST00000290573.6,ENST00000409174.1,ENST000004...</td>\n",
       "      <td>5250.08</td>\n",
       "      <td>5078.88</td>\n",
       "      <td>920.95</td>\n",
       "      <td>6.53</td>\n",
       "      <td>12.32</td>\n",
       "      <td>920.91</td>\n",
       "      <td>0.29</td>\n",
       "      <td>6.44</td>\n",
       "      <td>12.54</td>\n",
       "      <td>5.750730</td>\n",
       "      <td>7.220230</td>\n",
       "      <td>0.038445</td>\n",
       "      <td>11.179300</td>\n",
       "      <td>14.039000</td>\n",
       "      <td>0.038428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10999</th>\n",
       "      <td>ENSG00000159403.16</td>\n",
       "      <td>ENST00000535233.6,ENST00000536053.6,ENST000005...</td>\n",
       "      <td>2664.70</td>\n",
       "      <td>2493.49</td>\n",
       "      <td>84.00</td>\n",
       "      <td>1.21</td>\n",
       "      <td>2.29</td>\n",
       "      <td>83.63</td>\n",
       "      <td>0.66</td>\n",
       "      <td>2.60</td>\n",
       "      <td>5.05</td>\n",
       "      <td>1.830850</td>\n",
       "      <td>3.407680</td>\n",
       "      <td>0.105393</td>\n",
       "      <td>3.585890</td>\n",
       "      <td>6.654870</td>\n",
       "      <td>0.105349</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id                                   transcript_id(s)  \\\n",
       "10000   ENSG00000151006.7  ENST00000280606.6,ENST00000486499.1,ENST000004...   \n",
       "10001  ENSG00000151012.13                ENST00000280612.9,ENST00000509248.1   \n",
       "10002   ENSG00000151014.5  ENST00000280614.3,ENST00000515616.1,ENST000006...   \n",
       "10003  ENSG00000151023.16  ENST00000331161.8,ENST00000376363.5,ENST000004...   \n",
       "10004  ENSG00000151025.10  ENST00000376351.3,ENST00000482641.2,ENST000004...   \n",
       "...                   ...                                                ...   \n",
       "10995   ENSG00000159387.7                ENST00000290552.7,ENST00000558315.1   \n",
       "10996   ENSG00000159388.5                ENST00000290551.4,ENST00000475157.1   \n",
       "10997  ENSG00000159398.15  ENST00000290567.13,ENST00000319165.13,ENST0000...   \n",
       "10998   ENSG00000159399.9  ENST00000290573.6,ENST00000409174.1,ENST000004...   \n",
       "10999  ENSG00000159403.16  ENST00000535233.6,ENST00000536053.6,ENST000005...   \n",
       "\n",
       "        length  effective_length  expected_count   TPM   FPKM  \\\n",
       "10000  4510.90           4339.68          124.04  1.03   1.94   \n",
       "10001  9249.21           9077.99         1848.00  7.33  13.83   \n",
       "10002  1856.43           1685.21          115.00  2.46   4.64   \n",
       "10003  1318.00           1146.78            2.00  0.06   0.12   \n",
       "10004  4853.25           4682.04            0.00  0.00   0.00   \n",
       "...        ...               ...             ...   ...    ...   \n",
       "10995  1989.00           1817.78            0.00  0.00   0.00   \n",
       "10996  2579.06           2407.84          286.00  4.28   8.07   \n",
       "10997  1766.22           1595.01            0.00  0.00   0.00   \n",
       "10998  5250.08           5078.88          920.95  6.53  12.32   \n",
       "10999  2664.70           2493.49           84.00  1.21   2.29   \n",
       "\n",
       "       posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \\\n",
       "10000                134.12                                   6.64     1.21   \n",
       "10001               1848.00                                   0.00     7.14   \n",
       "10002                115.00                                   0.00     2.63   \n",
       "10003                  2.00                                   0.00     0.21   \n",
       "10004                  0.00                                   0.00     0.08   \n",
       "...                     ...                                    ...      ...   \n",
       "10995                  0.00                                   0.00     0.06   \n",
       "10996                286.00                                   0.00     4.15   \n",
       "10997                  0.00                                   0.00     0.25   \n",
       "10998                920.91                                   0.29     6.44   \n",
       "10999                 83.63                                   0.66     2.60   \n",
       "\n",
       "       pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "10000      2.36            0.872066            1.566580   \n",
       "10001     13.88            6.625750            7.708410   \n",
       "10002      5.11            2.013440            3.298030   \n",
       "10003      0.41            0.048159            0.425282   \n",
       "10004      0.15            0.004778            0.197461   \n",
       "...         ...                 ...                 ...   \n",
       "10995      0.12            0.000902            0.167404   \n",
       "10996      8.07            3.643900            4.658430   \n",
       "10997      0.50            0.085282            0.469493   \n",
       "10998     12.54            5.750730            7.220230   \n",
       "10999      5.05            1.830850            3.407680   \n",
       "\n",
       "       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "10000                               0.102375             1.704870   \n",
       "10001                               0.025499            12.886400   \n",
       "10002                               0.083487             3.915400   \n",
       "10003                               0.317992             0.093792   \n",
       "10004                               0.489654             0.010009   \n",
       "...                                      ...                  ...   \n",
       "10995                               0.532145             0.001434   \n",
       "10996                               0.042490             7.082570   \n",
       "10997                               0.262764             0.157796   \n",
       "10998                               0.038445            11.179300   \n",
       "10999                               0.105393             3.585890   \n",
       "\n",
       "       FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \n",
       "10000             3.056580                                0.102409  \n",
       "10001            14.991900                                0.025553  \n",
       "10002             6.415060                                0.083588  \n",
       "10003             0.827459                                0.317884  \n",
       "10004             0.384865                                0.489683  \n",
       "...                    ...                                     ...  \n",
       "10995             0.325286                                0.532105  \n",
       "10996             9.055910                                0.042501  \n",
       "10997             0.905380                                0.262659  \n",
       "10998            14.039000                                0.038428  \n",
       "10999             6.654870                                0.105349  \n",
       "\n",
       "[1000 rows x 17 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_geneQuant_df[10000:11000]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Seems to be a combination of ensemble genes, spike-ins (calibrations) and tRNA (digits - i.e. 10904)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histone ChIP-Seq data - H3K9me3 and HeK27me3 modifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading histone mod data\n",
    "# H3K9me3\n",
    "histone_pval_K9_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF125NHB.bedGraph', sep='\\t', names=['chrRef', 'start', 'end', 'pval'])\n",
    "histone_peaks_K9_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF170GMN.bed', sep='\\t', header=None)\n",
    "\n",
    "# H3K27me3\n",
    "histone_pval_K27_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF896BFP.bedGraph', sep='\\t', names=['chrRef', 'start', 'end', 'pval'])\n",
    "histone_peaks_K27_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF841QVP.bed', sep='\\t', header=None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histone columns:\n",
    "- Chromosome \n",
    "- start\n",
    "- end\n",
    "- peak label?\n",
    "- \n",
    "\n",
    "peaks:\n",
    "- chromosome\n",
    "- start\n",
    "- end \n",
    "- replication pct? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>991782</td>\n",
       "      <td>992368</td>\n",
       "      <td>Peak_49693</td>\n",
       "      <td>167</td>\n",
       "      <td>.</td>\n",
       "      <td>5.23056</td>\n",
       "      <td>16.71326</td>\n",
       "      <td>14.64974</td>\n",
       "      <td>235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1770030</td>\n",
       "      <td>1770624</td>\n",
       "      <td>Peak_107498</td>\n",
       "      <td>98</td>\n",
       "      <td>.</td>\n",
       "      <td>3.79372</td>\n",
       "      <td>9.87359</td>\n",
       "      <td>8.21870</td>\n",
       "      <td>315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148025</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26667979</td>\n",
       "      <td>26668865</td>\n",
       "      <td>Peak_48031</td>\n",
       "      <td>172</td>\n",
       "      <td>.</td>\n",
       "      <td>4.20870</td>\n",
       "      <td>17.27292</td>\n",
       "      <td>15.19134</td>\n",
       "      <td>673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148026</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26668976</td>\n",
       "      <td>26671930</td>\n",
       "      <td>Peak_4</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>6.71736</td>\n",
       "      <td>200.44463</td>\n",
       "      <td>194.12437</td>\n",
       "      <td>2116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148027</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26672410</td>\n",
       "      <td>26672736</td>\n",
       "      <td>Peak_91129</td>\n",
       "      <td>112</td>\n",
       "      <td>.</td>\n",
       "      <td>3.60158</td>\n",
       "      <td>11.25807</td>\n",
       "      <td>9.52160</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148028</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56679663</td>\n",
       "      <td>56680112</td>\n",
       "      <td>Peak_60182</td>\n",
       "      <td>151</td>\n",
       "      <td>.</td>\n",
       "      <td>4.94783</td>\n",
       "      <td>15.15114</td>\n",
       "      <td>13.17521</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148029</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56683855</td>\n",
       "      <td>56684306</td>\n",
       "      <td>Peak_10462</td>\n",
       "      <td>321</td>\n",
       "      <td>.</td>\n",
       "      <td>6.92330</td>\n",
       "      <td>32.11600</td>\n",
       "      <td>29.30741</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148030 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2            3     4  5        6          7  \\\n",
       "0       chr1    788438    790667   Peak_33943   206  .  4.85493   20.69931   \n",
       "1       chr1    991782    992368   Peak_49693   167  .  5.23056   16.71326   \n",
       "2       chr1   1112984   1113962   Peak_13060   297  .  7.35106   29.75177   \n",
       "3       chr1   1194302   1195375    Peak_6971   362  .  8.27128   36.28568   \n",
       "4       chr1   1770030   1770624  Peak_107498    98  .  3.79372    9.87359   \n",
       "...      ...       ...       ...          ...   ... ..      ...        ...   \n",
       "148025  chrY  26667979  26668865   Peak_48031   172  .  4.20870   17.27292   \n",
       "148026  chrY  26668976  26671930       Peak_4  1000  .  6.71736  200.44463   \n",
       "148027  chrY  26672410  26672736   Peak_91129   112  .  3.60158   11.25807   \n",
       "148028  chrY  56679663  56680112   Peak_60182   151  .  4.94783   15.15114   \n",
       "148029  chrY  56683855  56684306   Peak_10462   321  .  6.92330   32.11600   \n",
       "\n",
       "                8     9  \n",
       "0        18.43085   635  \n",
       "1        14.64974   235  \n",
       "2        27.03791   471  \n",
       "3        33.31710   650  \n",
       "4         8.21870   315  \n",
       "...           ...   ...  \n",
       "148025   15.19134   673  \n",
       "148026  194.12437  2116  \n",
       "148027    9.52160    94  \n",
       "148028   13.17521   124  \n",
       "148029   29.30741   227  \n",
       "\n",
       "[148030 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrRef</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>9944</td>\n",
       "      <td>0.00100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9944</td>\n",
       "      <td>9947</td>\n",
       "      <td>0.02630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9947</td>\n",
       "      <td>9950</td>\n",
       "      <td>0.13984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9950</td>\n",
       "      <td>9954</td>\n",
       "      <td>0.39194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9954</td>\n",
       "      <td>10150</td>\n",
       "      <td>0.79457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263757</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>17991</td>\n",
       "      <td>33769</td>\n",
       "      <td>0.00100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263758</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>33769</td>\n",
       "      <td>33984</td>\n",
       "      <td>0.02630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263759</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>33984</td>\n",
       "      <td>34006</td>\n",
       "      <td>0.13984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263760</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>34006</td>\n",
       "      <td>34212</td>\n",
       "      <td>0.02630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263761</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>34212</td>\n",
       "      <td>34224</td>\n",
       "      <td>0.00100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251263762 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           chrRef  start    end     pval\n",
       "0                            chr1      0   9944  0.00100\n",
       "1                            chr1   9944   9947  0.02630\n",
       "2                            chr1   9947   9950  0.13984\n",
       "3                            chr1   9950   9954  0.39194\n",
       "4                            chr1   9954  10150  0.79457\n",
       "...                           ...    ...    ...      ...\n",
       "251263757  chrY_KI270740v1_random  17991  33769  0.00100\n",
       "251263758  chrY_KI270740v1_random  33769  33984  0.02630\n",
       "251263759  chrY_KI270740v1_random  33984  34006  0.13984\n",
       "251263760  chrY_KI270740v1_random  34006  34212  0.02630\n",
       "251263761  chrY_KI270740v1_random  34212  34224  0.00100\n",
       "\n",
       "[251263762 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100236060</td>\n",
       "      <td>100236175</td>\n",
       "      <td>Peak_91116</td>\n",
       "      <td>33</td>\n",
       "      <td>.</td>\n",
       "      <td>3.38068</td>\n",
       "      <td>3.39518</td>\n",
       "      <td>0.84457</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100538099</td>\n",
       "      <td>100538408</td>\n",
       "      <td>Peak_9577</td>\n",
       "      <td>65</td>\n",
       "      <td>.</td>\n",
       "      <td>5.03472</td>\n",
       "      <td>6.59658</td>\n",
       "      <td>2.72930</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101135095</td>\n",
       "      <td>101135210</td>\n",
       "      <td>Peak_97001</td>\n",
       "      <td>32</td>\n",
       "      <td>.</td>\n",
       "      <td>3.29575</td>\n",
       "      <td>3.26597</td>\n",
       "      <td>0.83925</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101236811</td>\n",
       "      <td>101237102</td>\n",
       "      <td>Peak_17149</td>\n",
       "      <td>61</td>\n",
       "      <td>.</td>\n",
       "      <td>4.77464</td>\n",
       "      <td>6.16004</td>\n",
       "      <td>2.57973</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101237336</td>\n",
       "      <td>101237500</td>\n",
       "      <td>Peak_87875</td>\n",
       "      <td>36</td>\n",
       "      <td>.</td>\n",
       "      <td>3.27881</td>\n",
       "      <td>3.62443</td>\n",
       "      <td>1.03439</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51278</th>\n",
       "      <td>chrX</td>\n",
       "      <td>848527</td>\n",
       "      <td>848652</td>\n",
       "      <td>Peak_65447</td>\n",
       "      <td>43</td>\n",
       "      <td>.</td>\n",
       "      <td>3.93203</td>\n",
       "      <td>4.39984</td>\n",
       "      <td>1.48232</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51279</th>\n",
       "      <td>chrX</td>\n",
       "      <td>8742048</td>\n",
       "      <td>8742163</td>\n",
       "      <td>Peak_65448</td>\n",
       "      <td>43</td>\n",
       "      <td>.</td>\n",
       "      <td>3.93203</td>\n",
       "      <td>4.39984</td>\n",
       "      <td>1.48232</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51280</th>\n",
       "      <td>chrX</td>\n",
       "      <td>915821</td>\n",
       "      <td>915961</td>\n",
       "      <td>Peak_5605</td>\n",
       "      <td>77</td>\n",
       "      <td>.</td>\n",
       "      <td>6.13741</td>\n",
       "      <td>7.77377</td>\n",
       "      <td>3.56466</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51281</th>\n",
       "      <td>chrX</td>\n",
       "      <td>9295740</td>\n",
       "      <td>9295902</td>\n",
       "      <td>Peak_31774</td>\n",
       "      <td>54</td>\n",
       "      <td>.</td>\n",
       "      <td>4.48337</td>\n",
       "      <td>5.46977</td>\n",
       "      <td>2.10106</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51282</th>\n",
       "      <td>chrX</td>\n",
       "      <td>966389</td>\n",
       "      <td>966573</td>\n",
       "      <td>Peak_48390</td>\n",
       "      <td>49</td>\n",
       "      <td>.</td>\n",
       "      <td>4.47639</td>\n",
       "      <td>4.98931</td>\n",
       "      <td>1.86485</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51283 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          1          2           3   4  5        6        7  \\\n",
       "0      chr1  100236060  100236175  Peak_91116  33  .  3.38068  3.39518   \n",
       "1      chr1  100538099  100538408   Peak_9577  65  .  5.03472  6.59658   \n",
       "2      chr1  101135095  101135210  Peak_97001  32  .  3.29575  3.26597   \n",
       "3      chr1  101236811  101237102  Peak_17149  61  .  4.77464  6.16004   \n",
       "4      chr1  101237336  101237500  Peak_87875  36  .  3.27881  3.62443   \n",
       "...     ...        ...        ...         ...  .. ..      ...      ...   \n",
       "51278  chrX     848527     848652  Peak_65447  43  .  3.93203  4.39984   \n",
       "51279  chrX    8742048    8742163  Peak_65448  43  .  3.93203  4.39984   \n",
       "51280  chrX     915821     915961   Peak_5605  77  .  6.13741  7.77377   \n",
       "51281  chrX    9295740    9295902  Peak_31774  54  .  4.48337  5.46977   \n",
       "51282  chrX     966389     966573  Peak_48390  49  .  4.47639  4.98931   \n",
       "\n",
       "             8    9  \n",
       "0      0.84457   88  \n",
       "1      2.72930   79  \n",
       "2      0.83925   47  \n",
       "3      2.57973  107  \n",
       "4      1.03439   55  \n",
       "...        ...  ...  \n",
       "51278  1.48232   36  \n",
       "51279  1.48232   26  \n",
       "51280  3.56466   85  \n",
       "51281  2.10106   87  \n",
       "51282  1.86485  111  \n",
       "\n",
       "[51283 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrRef</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>10071</td>\n",
       "      <td>0.30592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10071</td>\n",
       "      <td>10100</td>\n",
       "      <td>0.82513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10100</td>\n",
       "      <td>10127</td>\n",
       "      <td>0.50508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10127</td>\n",
       "      <td>10186</td>\n",
       "      <td>0.18005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10186</td>\n",
       "      <td>10215</td>\n",
       "      <td>0.04769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656086</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>29725</td>\n",
       "      <td>31121</td>\n",
       "      <td>0.30592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656087</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>31121</td>\n",
       "      <td>31236</td>\n",
       "      <td>0.82513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656088</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>31236</td>\n",
       "      <td>36460</td>\n",
       "      <td>0.30592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656089</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>36460</td>\n",
       "      <td>36487</td>\n",
       "      <td>0.82513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656090</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>36487</td>\n",
       "      <td>36575</td>\n",
       "      <td>0.50508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59656091 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          chrRef  start    end     pval\n",
       "0                           chr1      0  10071  0.30592\n",
       "1                           chr1  10071  10100  0.82513\n",
       "2                           chr1  10100  10127  0.50508\n",
       "3                           chr1  10127  10186  0.18005\n",
       "4                           chr1  10186  10215  0.04769\n",
       "...                          ...    ...    ...      ...\n",
       "59656086  chrY_KI270740v1_random  29725  31121  0.30592\n",
       "59656087  chrY_KI270740v1_random  31121  31236  0.82513\n",
       "59656088  chrY_KI270740v1_random  31236  36460  0.30592\n",
       "59656089  chrY_KI270740v1_random  36460  36487  0.82513\n",
       "59656090  chrY_KI270740v1_random  36487  36575  0.50508\n",
       "\n",
       "[59656091 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(histone_peaks_K9_df)\n",
    "display(histone_pval_K9_df)\n",
    "\n",
    "display(histone_peaks_K27_df)\n",
    "display(histone_pval_K27_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualising\n",
    "\n",
    "To begin, we utilised the HepG2 WGBS methylation state at CPG, RNA-Seq Gene Quantifications, and Histone ChIP-Seq pseudoreplicated peak files (loading them in their original forms). \n",
    "- ENCFF690FNR (first 9 columns only): WGBS\n",
    "- ENCFF194TSC: RNA minus\n",
    "- ENCFF792BNC: RNA plus\n",
    "- ENCFF170GMN: Hist H3K9me3 pseudorep\n",
    "- ENCFF505LOU: Hist H3K27me3 pseudorep"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "WGBS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combining RNA gene quantification file with V29 gene annotations?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histone ChIP-Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intervals\n",
    "First, annotate data with genes and aggregate at the gene level.\n",
    "\n",
    "Compute the sum of modifications for an entire gene plus its transcription start site. An example model might have the signature: (sum DNA methylation, sum H3K9me3, ...etc) -> (gene expression level). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene annotations for WBGS data - can be applied directly to it using chr, start and end\n",
    "# Gene annotations for RNA data - not sure how to approach this\n",
    "# Gene annotations for hist CS data - do I used pseudo or p-val? annotate both using bedtools intersect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>startChr</th>\n",
       "      <th>endChr</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>startThick</th>\n",
       "      <th>endThick</th>\n",
       "      <th>rgb</th>\n",
       "      <th>reads</th>\n",
       "      <th>pctMe</th>\n",
       "      <th>refGen</th>\n",
       "      <th>sampGen</th>\n",
       "      <th>qualScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CG</td>\n",
       "      <td>MG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>155,255,0</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>CG</td>\n",
       "      <td>CK</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51350653</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35645</td>\n",
       "      <td>35646</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>35645</td>\n",
       "      <td>35646</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51350654</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35646</td>\n",
       "      <td>35647</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>35646</td>\n",
       "      <td>35647</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51350655</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35685</td>\n",
       "      <td>35686</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>35685</td>\n",
       "      <td>35686</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51350656</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35686</td>\n",
       "      <td>35687</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>35686</td>\n",
       "      <td>35687</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51350657</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35862</td>\n",
       "      <td>35863</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>35862</td>\n",
       "      <td>35863</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>3</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51350658 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          refChr  startChr  endChr  name  score strand  \\\n",
       "0                           chr1     10468   10469     1      1      +   \n",
       "1                           chr1     10469   10470     1      4      -   \n",
       "2                           chr1     10470   10471     1      1      +   \n",
       "3                           chr1     10471   10472     1      2      -   \n",
       "4                           chr1     10483   10484     1      1      +   \n",
       "...                          ...       ...     ...   ...    ...    ...   \n",
       "51350653  chrY_KI270740v1_random     35645   35646     1      1      +   \n",
       "51350654  chrY_KI270740v1_random     35646   35647     1      1      -   \n",
       "51350655  chrY_KI270740v1_random     35685   35686     1      1      +   \n",
       "51350656  chrY_KI270740v1_random     35686   35687     1      1      -   \n",
       "51350657  chrY_KI270740v1_random     35862   35863     1      3      -   \n",
       "\n",
       "          startThick  endThick        rgb  reads  pctMe refGen sampGen  \\\n",
       "0              10468     10469    0,255,0      1      0     CG      MG   \n",
       "1              10469     10470  155,255,0      4     25     CG      CK   \n",
       "2              10470     10471    255,0,0      1    100     CG      CG   \n",
       "3              10471     10472    255,0,0      2    100     CG      CG   \n",
       "4              10483     10484    255,0,0      1    100     CG      CG   \n",
       "...              ...       ...        ...    ...    ...    ...     ...   \n",
       "51350653       35645     35646    0,255,0      1      0     CG      CG   \n",
       "51350654       35646     35647    0,255,0      1      0     CG      CG   \n",
       "51350655       35685     35686    0,255,0      1      0     CG      CG   \n",
       "51350656       35686     35687    0,255,0      1      0     CG      CG   \n",
       "51350657       35862     35863    255,0,0      3    100     CG      CG   \n",
       "\n",
       "          qualScore  \n",
       "0                 3  \n",
       "1                 5  \n",
       "2                11  \n",
       "3                 6  \n",
       "4                 4  \n",
       "...             ...  \n",
       "51350653          3  \n",
       "51350654          3  \n",
       "51350655          3  \n",
       "51350656          3  \n",
       "51350657          6  \n",
       "\n",
       "[51350658 rows x 14 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>startChr</th>\n",
       "      <th>endChr</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>startThick</th>\n",
       "      <th>endThick</th>\n",
       "      <th>rgb</th>\n",
       "      <th>reads</th>\n",
       "      <th>pctMe</th>\n",
       "      <th>refGen</th>\n",
       "      <th>sampGen</th>\n",
       "      <th>qualScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>155,255,0</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>CG</td>\n",
       "      <td>CK</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10484</td>\n",
       "      <td>10485</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>10484</td>\n",
       "      <td>10485</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10488</td>\n",
       "      <td>10489</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10488</td>\n",
       "      <td>10489</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10489</td>\n",
       "      <td>10490</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>10489</td>\n",
       "      <td>10490</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10492</td>\n",
       "      <td>10493</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10492</td>\n",
       "      <td>10493</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>MG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10493</td>\n",
       "      <td>10494</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>10493</td>\n",
       "      <td>10494</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CG</td>\n",
       "      <td>CK</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refChr  startChr  endChr  name  score strand  startThick  endThick  \\\n",
       "1   chr1     10469   10470     1      4      -       10469     10470   \n",
       "2   chr1     10470   10471     1      1      +       10470     10471   \n",
       "3   chr1     10471   10472     1      2      -       10471     10472   \n",
       "4   chr1     10483   10484     1      1      +       10483     10484   \n",
       "5   chr1     10484   10485     1      1      -       10484     10485   \n",
       "6   chr1     10488   10489     1      1      +       10488     10489   \n",
       "7   chr1     10489   10490     1      1      -       10489     10490   \n",
       "8   chr1     10492   10493     1      1      +       10492     10493   \n",
       "9   chr1     10493   10494     1      1      -       10493     10494   \n",
       "\n",
       "         rgb  reads  pctMe refGen sampGen  qualScore  \n",
       "1  155,255,0      4     25     CG      CK          5  \n",
       "2    255,0,0      1    100     CG      CG         11  \n",
       "3    255,0,0      2    100     CG      CG          6  \n",
       "4    255,0,0      1    100     CG      CG          4  \n",
       "5    255,0,0      1    100     CG      CG          4  \n",
       "6    255,0,0      1    100     CG      CG          4  \n",
       "7    255,0,0      1    100     CG      CG          4  \n",
       "8    255,0,0      1    100     CG      MG          3  \n",
       "9    0,255,0      1      0     CG      CK          3  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display(WGBS_df)\n",
    "WGBS_df.iloc[1:10,0:15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "ParserError",
     "evalue": "Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mParserError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m#loading gene annotated wgbs data\u001b[39;00m\n\u001b[1;32m----> 2\u001b[0m WGBS_plus_df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mHepG2_data/HepG2_DNAm/processed/ENCFF884UAE.bedGraph\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnames\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrefChr\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mstart\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mend\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpctMe\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m      3\u001b[0m WGBS_minus_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHepG2_data/HepG2_DNAm/processed/ENCFF464CWC.bedGraph\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, names\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrefChr\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mstart\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mend\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpctMe\u001b[39m\u001b[38;5;124m'\u001b[39m], header\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m)\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\util\\_decorators.py:211\u001b[0m, in \u001b[0;36mdeprecate_kwarg.<locals>._deprecate_kwarg.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    209\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    210\u001b[0m         kwargs[new_arg_name] \u001b[38;5;241m=\u001b[39m new_arg_value\n\u001b[1;32m--> 211\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\util\\_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    325\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[0;32m    326\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[0;32m    327\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[0;32m    328\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[0;32m    329\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[0;32m    330\u001b[0m     )\n\u001b[1;32m--> 331\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:950\u001b[0m, in \u001b[0;36mread_csv\u001b[1;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[0;32m    935\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[0;32m    936\u001b[0m     dialect,\n\u001b[0;32m    937\u001b[0m     delimiter,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    946\u001b[0m     defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[0;32m    947\u001b[0m )\n\u001b[0;32m    948\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[1;32m--> 950\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:611\u001b[0m, in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    608\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n\u001b[0;32m    610\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[1;32m--> 611\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:1778\u001b[0m, in \u001b[0;36mTextFileReader.read\u001b[1;34m(self, nrows)\u001b[0m\n\u001b[0;32m   1771\u001b[0m nrows \u001b[38;5;241m=\u001b[39m validate_integer(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnrows\u001b[39m\u001b[38;5;124m\"\u001b[39m, nrows)\n\u001b[0;32m   1772\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1773\u001b[0m     \u001b[38;5;66;03m# error: \"ParserBase\" has no attribute \"read\"\u001b[39;00m\n\u001b[0;32m   1774\u001b[0m     (\n\u001b[0;32m   1775\u001b[0m         index,\n\u001b[0;32m   1776\u001b[0m         columns,\n\u001b[0;32m   1777\u001b[0m         col_dict,\n\u001b[1;32m-> 1778\u001b[0m     ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[attr-defined]\u001b[39;49;00m\n\u001b[0;32m   1779\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\n\u001b[0;32m   1780\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1781\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[0;32m   1782\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\c_parser_wrapper.py:230\u001b[0m, in \u001b[0;36mCParserWrapper.read\u001b[1;34m(self, nrows)\u001b[0m\n\u001b[0;32m    228\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    229\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlow_memory:\n\u001b[1;32m--> 230\u001b[0m         chunks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_reader\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_low_memory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    231\u001b[0m         \u001b[38;5;66;03m# destructive to chunks\u001b[39;00m\n\u001b[0;32m    232\u001b[0m         data \u001b[38;5;241m=\u001b[39m _concatenate_chunks(chunks)\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:808\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.read_low_memory\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:866\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._read_rows\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:852\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\julia\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\_libs\\parsers.pyx:1973\u001b[0m, in \u001b[0;36mpandas._libs.parsers.raise_parser_error\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mParserError\u001b[0m: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'."
     ]
    }
   ],
   "source": [
    "#loading gene annotated wgbs data\n",
    "WGBS_plus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF884UAE.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'pctMe'], header=None)\n",
    "WGBS_minus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF464CWC.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'pctMe'], header=None)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>chr1</td>\n",
       "      <td>805670</td>\n",
       "      <td>805671</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>805670</td>\n",
       "      <td>805671</td>\n",
       "      <td>255,55,0</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>chr1</td>\n",
       "      <td>805670</td>\n",
       "      <td>805671</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>805670</td>\n",
       "      <td>805671</td>\n",
       "      <td>255,55,0</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>594308</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>chr1</td>\n",
       "      <td>805670</td>\n",
       "      <td>805671</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>805670</td>\n",
       "      <td>805671</td>\n",
       "      <td>255,55,0</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778770</td>\n",
       "      <td>810060</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>chr1</td>\n",
       "      <td>805723</td>\n",
       "      <td>805724</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>805723</td>\n",
       "      <td>805724</td>\n",
       "      <td>105,255,0</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>804799</td>\n",
       "      <td>807465</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>chr1</td>\n",
       "      <td>805723</td>\n",
       "      <td>805724</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>805723</td>\n",
       "      <td>805724</td>\n",
       "      <td>105,255,0</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>804932</td>\n",
       "      <td>810060</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        0       1       2   3   4  5       6       7          8   9   ...  13  \\\n",
       "0     chr1   14709   14710   1   1  +   14709   14710    255,0,0   1  ...   2   \n",
       "1     chr1   14709   14710   1   1  +   14709   14710    255,0,0   1  ...   2   \n",
       "2     chr1   14715   14716   1   1  +   14715   14716    255,0,0   1  ...   2   \n",
       "3     chr1   14715   14716   1   1  +   14715   14716    255,0,0   1  ...   2   \n",
       "4     chr1   14740   14741   1   1  +   14740   14741    255,0,0   1  ...   2   \n",
       "...    ...     ...     ...  ..  .. ..     ...     ...        ...  ..  ...  ..   \n",
       "9995  chr1  805670  805671   1  12  -  805670  805671   255,55,0  12  ...  24   \n",
       "9996  chr1  805670  805671   1  12  -  805670  805671   255,55,0  12  ...  24   \n",
       "9997  chr1  805670  805671   1  12  -  805670  805671   255,55,0  12  ...  24   \n",
       "9998  chr1  805723  805724   1   6  -  805723  805724  105,255,0   6  ...   3   \n",
       "9999  chr1  805723  805724   1   6  -  805723  805724  105,255,0   6  ...   3   \n",
       "\n",
       "        14      15          16      17      18 19  20  21  \\\n",
       "0     chr1  HAVANA        gene   14404   29570  .   -   .   \n",
       "1     chr1  HAVANA  transcript   14404   29570  .   -   .   \n",
       "2     chr1  HAVANA        gene   14404   29570  .   -   .   \n",
       "3     chr1  HAVANA  transcript   14404   29570  .   -   .   \n",
       "4     chr1  HAVANA        gene   14404   29570  .   -   .   \n",
       "...    ...     ...         ...     ...     ... ..  ..  ..   \n",
       "9995  chr1  HAVANA        gene  586071  827796  .   -   .   \n",
       "9996  chr1  HAVANA  transcript  594308  827796  .   -   .   \n",
       "9997  chr1  HAVANA        gene  778770  810060  .   +   .   \n",
       "9998  chr1  HAVANA  transcript  804799  807465  .   +   .   \n",
       "9999  chr1  HAVANA  transcript  804932  810060  .   +   .   \n",
       "\n",
       "                                                     22  \n",
       "0     gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "1     gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "2     gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "3     gene_id \"ENSG00000227232.5\"; transcript_id \"EN...  \n",
       "4     gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "...                                                 ...  \n",
       "9995  gene_id \"ENSG00000230021.10\"; gene_type \"trans...  \n",
       "9996  gene_id \"ENSG00000230021.10\"; transcript_id \"E...  \n",
       "9997  gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...  \n",
       "9998  gene_id \"ENSG00000237491.8\"; transcript_id \"EN...  \n",
       "9999  gene_id \"ENSG00000237491.8\"; transcript_id \"EN...  \n",
       "\n",
       "[10000 rows x 23 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0      1      2   3   4  5      6      7        8   9    10  11  12  13  \\\n",
       "0  chr1  14709  14710   1   1  +  14709  14710  255,0,0   1  100  CG  CG   2   \n",
       "1  chr1  14709  14710   1   1  +  14709  14710  255,0,0   1  100  CG  CG   2   \n",
       "2  chr1  14715  14716   1   1  +  14715  14716  255,0,0   1  100  CG  CG   2   \n",
       "3  chr1  14715  14716   1   1  +  14715  14716  255,0,0   1  100  CG  CG   2   \n",
       "4  chr1  14740  14741   1   1  +  14740  14741  255,0,0   1  100  CG  CG   2   \n",
       "5  chr1  14740  14741   1   1  +  14740  14741  255,0,0   1  100  CG  CG   2   \n",
       "6  chr1  14746  14747   1   1  +  14746  14747  255,0,0   1  100  CG  CG   2   \n",
       "7  chr1  14746  14747   1   1  +  14746  14747  255,0,0   1  100  CG  CG   2   \n",
       "8  chr1  14772  14773   1   1  +  14772  14773  255,0,0   1  100  CG  CG   2   \n",
       "9  chr1  14772  14773   1   1  +  14772  14773  255,0,0   1  100  CG  CG   2   \n",
       "\n",
       "     14  \n",
       "0  chr1  \n",
       "1  chr1  \n",
       "2  chr1  \n",
       "3  chr1  \n",
       "4  chr1  \n",
       "5  chr1  \n",
       "6  chr1  \n",
       "7  chr1  \n",
       "8  chr1  \n",
       "9  chr1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>155,255,0</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10484</td>\n",
       "      <td>10485</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>10484</td>\n",
       "      <td>10485</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10488</td>\n",
       "      <td>10489</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10488</td>\n",
       "      <td>10489</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10489</td>\n",
       "      <td>10490</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>10489</td>\n",
       "      <td>10490</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10492</td>\n",
       "      <td>10493</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10492</td>\n",
       "      <td>10493</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10493</td>\n",
       "      <td>10494</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>10493</td>\n",
       "      <td>10494</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     0      1      2   3   4  5      6      7          8   9   ...  13 14 15  \\\n",
       "0  chr1  10468  10469   1   1  +  10468  10469    0,255,0   1  ...   3  .  .   \n",
       "1  chr1  10469  10470   1   4  -  10469  10470  155,255,0   4  ...   5  .  .   \n",
       "2  chr1  10470  10471   1   1  +  10470  10471    255,0,0   1  ...  11  .  .   \n",
       "3  chr1  10471  10472   1   2  -  10471  10472    255,0,0   2  ...   6  .  .   \n",
       "4  chr1  10483  10484   1   1  +  10483  10484    255,0,0   1  ...   4  .  .   \n",
       "5  chr1  10484  10485   1   1  -  10484  10485    255,0,0   1  ...   4  .  .   \n",
       "6  chr1  10488  10489   1   1  +  10488  10489    255,0,0   1  ...   4  .  .   \n",
       "7  chr1  10489  10490   1   1  -  10489  10490    255,0,0   1  ...   4  .  .   \n",
       "8  chr1  10492  10493   1   1  +  10492  10493    255,0,0   1  ...   3  .  .   \n",
       "9  chr1  10493  10494   1   1  -  10493  10494    0,255,0   1  ...   3  .  .   \n",
       "\n",
       "   16 17 18 19  20  21 22  \n",
       "0   . -1 -1  .   .   .  .  \n",
       "1   . -1 -1  .   .   .  .  \n",
       "2   . -1 -1  .   .   .  .  \n",
       "3   . -1 -1  .   .   .  .  \n",
       "4   . -1 -1  .   .   .  .  \n",
       "5   . -1 -1  .   .   .  .  \n",
       "6   . -1 -1  .   .   .  .  \n",
       "7   . -1 -1  .   .   .  .  \n",
       "8   . -1 -1  .   .   .  .  \n",
       "9   . -1 -1  .   .   .  .  \n",
       "\n",
       "[10 rows x 23 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(wgbs_annotated_df)\n",
    "display(wgbs_annotated_df.iloc[0:10, 0:15])\n",
    "#display(wgbs_annotated_df2.iloc[0:10])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>5</th>\n",
       "      <th>10</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>+</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>-</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>+</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10484</td>\n",
       "      <td>10485</td>\n",
       "      <td>-</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10488</td>\n",
       "      <td>10489</td>\n",
       "      <td>+</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10489</td>\n",
       "      <td>10490</td>\n",
       "      <td>-</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10492</td>\n",
       "      <td>10493</td>\n",
       "      <td>+</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10493</td>\n",
       "      <td>10494</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0      1      2  5    10 22\n",
       "0  chr1  10468  10469  +    0  .\n",
       "1  chr1  10469  10470  -   25  .\n",
       "2  chr1  10470  10471  +  100  .\n",
       "3  chr1  10471  10472  -  100  .\n",
       "4  chr1  10483  10484  +  100  .\n",
       "5  chr1  10484  10485  -  100  .\n",
       "6  chr1  10488  10489  +  100  .\n",
       "7  chr1  10489  10490  -  100  .\n",
       "8  chr1  10492  10493  +  100  .\n",
       "9  chr1  10493  10494  -    0  ."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "wgbs_annotated_df = pd.read_csv('HepG2_data/HepG2_DNAm/annotated/wgbs_with_gene_annotations_loj.bed', sep='\\t', header=None, usecols=[0, 1, 2, 5, 10, 22] )\n",
    "display(wgbs_annotated_df.iloc[0:10,0:15])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pattern = r'gene_id\\s*\"([^\"]+)\"'\n",
    "\n",
    "wgbs_annotated_df.loc[:, 'GeneID'] = wgbs_annotated_df.iloc[:, -1].str.extract(pattern)\n",
    "wgbs_annotated_df.loc[:, 'GeneID'] = wgbs_annotated_df.iloc[:, -1].str.extract(pattern)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>5</th>\n",
       "      <th>10</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0      1      2  5   10 22\n",
       "0  chr1  10468  10469  +   0  ."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\julia\\AppData\\Local\\Temp\\ipykernel_9980\\2183389383.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  test.loc[:, 'GeneID'] = test.iloc[:, -1].str.extract(pattern)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>5</th>\n",
       "      <th>10</th>\n",
       "      <th>22</th>\n",
       "      <th>GeneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2  5  10 22 GeneID\n",
       "0  chr1  10468  10469  +   0  .    NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    .\n",
      "Name: 22, dtype: object\n"
     ]
    }
   ],
   "source": [
    "test = wgbs_annotated_df.iloc[0:1]\n",
    "display(test)\n",
    "test.loc[:, 'GeneID'] = test.iloc[:, -1].str.extract(pattern)\n",
    "display(test)\n",
    "test2 = wgbs_annotated_df.iloc[0:1]\n",
    "print(test2[22])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>5</th>\n",
       "      <th>10</th>\n",
       "      <th>22</th>\n",
       "      <th>GeneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>-</td>\n",
       "      <td>25</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>+</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>-</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>+</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156345308</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35645</td>\n",
       "      <td>35646</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156345309</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35646</td>\n",
       "      <td>35647</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156345310</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35685</td>\n",
       "      <td>35686</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156345311</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35686</td>\n",
       "      <td>35687</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156345312</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>35862</td>\n",
       "      <td>35863</td>\n",
       "      <td>-</td>\n",
       "      <td>100</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156345313 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                0      1      2  5   10 22 GeneID\n",
       "0                            chr1  10468  10469  +    0  .    NaN\n",
       "1                            chr1  10469  10470  -   25  .    NaN\n",
       "2                            chr1  10470  10471  +  100  .    NaN\n",
       "3                            chr1  10471  10472  -  100  .    NaN\n",
       "4                            chr1  10483  10484  +  100  .    NaN\n",
       "...                           ...    ...    ... ..  ... ..    ...\n",
       "156345308  chrY_KI270740v1_random  35645  35646  +    0  .    NaN\n",
       "156345309  chrY_KI270740v1_random  35646  35647  -    0  .    NaN\n",
       "156345310  chrY_KI270740v1_random  35685  35686  +    0  .    NaN\n",
       "156345311  chrY_KI270740v1_random  35686  35687  -    0  .    NaN\n",
       "156345312  chrY_KI270740v1_random  35862  35863  -  100  .    NaN\n",
       "\n",
       "[156345313 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(wgbs_annotated_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate histone chipseq files\n",
    "h3k9me3_pp_annotated_df = pd.read_csv('HepG2_data/HepG2_histone/annotated/H3K9me3_pp_annotations.bed', sep='\\t', header=None, nrows=1000000)\n",
    "#h3k9me3_pval_annotated_df = pd.read_csv('HepG2_data/HepG2_histone/annotated/H3K9me3_pp_annotations.bed', sep='\\t', header=None, nrows=1000000)\n",
    "h3k27me3_pp_annotated_df = pd.read_csv('HepG2_data/HepG2_histone/annotated/H3K9me3_pp_annotations.bed', sep='\\t', header=None, nrows=1000000)\n",
    "#h3k27me3_pval_annotated_df = pd.read_csv('HepG2_data/HepG2_histone/annotated/H3K9me3_pp_annotations.bed', sep='\\t', header=None, nrows=1000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>991782</td>\n",
       "      <td>992368</td>\n",
       "      <td>Peak_49693</td>\n",
       "      <td>167</td>\n",
       "      <td>.</td>\n",
       "      <td>5.23056</td>\n",
       "      <td>16.71326</td>\n",
       "      <td>14.64974</td>\n",
       "      <td>235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1770030</td>\n",
       "      <td>1770624</td>\n",
       "      <td>Peak_107498</td>\n",
       "      <td>98</td>\n",
       "      <td>.</td>\n",
       "      <td>3.79372</td>\n",
       "      <td>9.87359</td>\n",
       "      <td>8.21870</td>\n",
       "      <td>315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1771153</td>\n",
       "      <td>1772290</td>\n",
       "      <td>Peak_9246</td>\n",
       "      <td>334</td>\n",
       "      <td>.</td>\n",
       "      <td>6.34875</td>\n",
       "      <td>33.41285</td>\n",
       "      <td>30.55444</td>\n",
       "      <td>225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2077026</td>\n",
       "      <td>2078865</td>\n",
       "      <td>Peak_2373</td>\n",
       "      <td>493</td>\n",
       "      <td>.</td>\n",
       "      <td>9.31909</td>\n",
       "      <td>49.38892</td>\n",
       "      <td>45.98370</td>\n",
       "      <td>1237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2422970</td>\n",
       "      <td>2423938</td>\n",
       "      <td>Peak_19118</td>\n",
       "      <td>260</td>\n",
       "      <td>.</td>\n",
       "      <td>6.78560</td>\n",
       "      <td>26.06830</td>\n",
       "      <td>23.51803</td>\n",
       "      <td>570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2429237</td>\n",
       "      <td>2430460</td>\n",
       "      <td>Peak_45478</td>\n",
       "      <td>175</td>\n",
       "      <td>.</td>\n",
       "      <td>5.37193</td>\n",
       "      <td>17.51166</td>\n",
       "      <td>15.40157</td>\n",
       "      <td>859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2488544</td>\n",
       "      <td>2489340</td>\n",
       "      <td>Peak_92748</td>\n",
       "      <td>109</td>\n",
       "      <td>.</td>\n",
       "      <td>3.80781</td>\n",
       "      <td>10.96670</td>\n",
       "      <td>9.23946</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1        2            3    4  5        6         7         8  \\\n",
       "0  chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "1  chr1   991782   992368   Peak_49693  167  .  5.23056  16.71326  14.64974   \n",
       "2  chr1  1112984  1113962   Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "3  chr1  1194302  1195375    Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "4  chr1  1770030  1770624  Peak_107498   98  .  3.79372   9.87359   8.21870   \n",
       "5  chr1  1771153  1772290    Peak_9246  334  .  6.34875  33.41285  30.55444   \n",
       "6  chr1  2077026  2078865    Peak_2373  493  .  9.31909  49.38892  45.98370   \n",
       "7  chr1  2422970  2423938   Peak_19118  260  .  6.78560  26.06830  23.51803   \n",
       "8  chr1  2429237  2430460   Peak_45478  175  .  5.37193  17.51166  15.40157   \n",
       "9  chr1  2488544  2489340   Peak_92748  109  .  3.80781  10.96670   9.23946   \n",
       "\n",
       "      9  \n",
       "0   635  \n",
       "1   235  \n",
       "2   471  \n",
       "3   650  \n",
       "4   315  \n",
       "5   225  \n",
       "6  1237  \n",
       "7   570  \n",
       "8   859  \n",
       "9   232  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100236060</td>\n",
       "      <td>100236175</td>\n",
       "      <td>Peak_91116</td>\n",
       "      <td>33</td>\n",
       "      <td>.</td>\n",
       "      <td>3.38068</td>\n",
       "      <td>3.39518</td>\n",
       "      <td>0.84457</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100538099</td>\n",
       "      <td>100538408</td>\n",
       "      <td>Peak_9577</td>\n",
       "      <td>65</td>\n",
       "      <td>.</td>\n",
       "      <td>5.03472</td>\n",
       "      <td>6.59658</td>\n",
       "      <td>2.72930</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101135095</td>\n",
       "      <td>101135210</td>\n",
       "      <td>Peak_97001</td>\n",
       "      <td>32</td>\n",
       "      <td>.</td>\n",
       "      <td>3.29575</td>\n",
       "      <td>3.26597</td>\n",
       "      <td>0.83925</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101236811</td>\n",
       "      <td>101237102</td>\n",
       "      <td>Peak_17149</td>\n",
       "      <td>61</td>\n",
       "      <td>.</td>\n",
       "      <td>4.77464</td>\n",
       "      <td>6.16004</td>\n",
       "      <td>2.57973</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101237336</td>\n",
       "      <td>101237500</td>\n",
       "      <td>Peak_87875</td>\n",
       "      <td>36</td>\n",
       "      <td>.</td>\n",
       "      <td>3.27881</td>\n",
       "      <td>3.62443</td>\n",
       "      <td>1.03439</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101240137</td>\n",
       "      <td>101240569</td>\n",
       "      <td>Peak_71693</td>\n",
       "      <td>40</td>\n",
       "      <td>.</td>\n",
       "      <td>3.59275</td>\n",
       "      <td>4.07966</td>\n",
       "      <td>1.32727</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101250307</td>\n",
       "      <td>101250422</td>\n",
       "      <td>Peak_71694</td>\n",
       "      <td>40</td>\n",
       "      <td>.</td>\n",
       "      <td>3.59275</td>\n",
       "      <td>4.07966</td>\n",
       "      <td>1.32727</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101314378</td>\n",
       "      <td>101314496</td>\n",
       "      <td>Peak_39327</td>\n",
       "      <td>50</td>\n",
       "      <td>.</td>\n",
       "      <td>4.22304</td>\n",
       "      <td>5.04547</td>\n",
       "      <td>1.91208</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10695837</td>\n",
       "      <td>10696061</td>\n",
       "      <td>Peak_34048</td>\n",
       "      <td>52</td>\n",
       "      <td>.</td>\n",
       "      <td>4.37074</td>\n",
       "      <td>5.28382</td>\n",
       "      <td>2.07027</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1070212</td>\n",
       "      <td>1070504</td>\n",
       "      <td>Peak_4701</td>\n",
       "      <td>78</td>\n",
       "      <td>.</td>\n",
       "      <td>5.93909</td>\n",
       "      <td>7.81863</td>\n",
       "      <td>3.59440</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2           3   4  5        6        7        8  \\\n",
       "0  chr1  100236060  100236175  Peak_91116  33  .  3.38068  3.39518  0.84457   \n",
       "1  chr1  100538099  100538408   Peak_9577  65  .  5.03472  6.59658  2.72930   \n",
       "2  chr1  101135095  101135210  Peak_97001  32  .  3.29575  3.26597  0.83925   \n",
       "3  chr1  101236811  101237102  Peak_17149  61  .  4.77464  6.16004  2.57973   \n",
       "4  chr1  101237336  101237500  Peak_87875  36  .  3.27881  3.62443  1.03439   \n",
       "5  chr1  101240137  101240569  Peak_71693  40  .  3.59275  4.07966  1.32727   \n",
       "6  chr1  101250307  101250422  Peak_71694  40  .  3.59275  4.07966  1.32727   \n",
       "7  chr1  101314378  101314496  Peak_39327  50  .  4.22304  5.04547  1.91208   \n",
       "8  chr1   10695837   10696061  Peak_34048  52  .  4.37074  5.28382  2.07027   \n",
       "9  chr1    1070212    1070504   Peak_4701  78  .  5.93909  7.81863  3.59440   \n",
       "\n",
       "     9  \n",
       "0   88  \n",
       "1   79  \n",
       "2   47  \n",
       "3  107  \n",
       "4   55  \n",
       "5   94  \n",
       "6   84  \n",
       "7   64  \n",
       "8  174  \n",
       "9   70  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>594308</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778770</td>\n",
       "      <td>810060</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>779056</td>\n",
       "      <td>804875</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1081818</td>\n",
       "      <td>1116361</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1081818</td>\n",
       "      <td>1116081</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1081823</td>\n",
       "      <td>1116098</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1173884</td>\n",
       "      <td>1197935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162571.13\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1173903</td>\n",
       "      <td>1197935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162571.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1173906</td>\n",
       "      <td>1197933</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162571.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0        1        2           3    4  5        6         7         8   \\\n",
       "0  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "1  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "2  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "3  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "4  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "5  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "6  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "7  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "8  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "9  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "\n",
       "    9     10       11          12       13       14 15 16 17  \\\n",
       "0  635  chr1   HAVANA        gene   586071   827796  .  -  .   \n",
       "1  635  chr1   HAVANA  transcript   594308   827796  .  -  .   \n",
       "2  635  chr1   HAVANA        gene   778770   810060  .  +  .   \n",
       "3  635  chr1   HAVANA  transcript   779056   804875  .  +  .   \n",
       "4  471  chr1   HAVANA        gene  1081818  1116361  .  -  .   \n",
       "5  471  chr1   HAVANA  transcript  1081818  1116081  .  -  .   \n",
       "6  471  chr1   HAVANA  transcript  1081823  1116098  .  -  .   \n",
       "7  650  chr1   HAVANA        gene  1173884  1197935  .  +  .   \n",
       "8  650  chr1   HAVANA  transcript  1173903  1197935  .  +  .   \n",
       "9  650  chr1  ENSEMBL  transcript  1173906  1197933  .  +  .   \n",
       "\n",
       "                                                  18  \n",
       "0  gene_id \"ENSG00000230021.10\"; gene_type \"trans...  \n",
       "1  gene_id \"ENSG00000230021.10\"; transcript_id \"E...  \n",
       "2  gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...  \n",
       "3  gene_id \"ENSG00000237491.8\"; transcript_id \"EN...  \n",
       "4  gene_id \"ENSG00000131591.17\"; gene_type \"prote...  \n",
       "5  gene_id \"ENSG00000131591.17\"; transcript_id \"E...  \n",
       "6  gene_id \"ENSG00000131591.17\"; transcript_id \"E...  \n",
       "7  gene_id \"ENSG00000162571.13\"; gene_type \"prote...  \n",
       "8  gene_id \"ENSG00000162571.13\"; transcript_id \"E...  \n",
       "9  gene_id \"ENSG00000162571.13\"; transcript_id \"E...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>594308</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778770</td>\n",
       "      <td>810060</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>779056</td>\n",
       "      <td>804875</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1081818</td>\n",
       "      <td>1116361</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1081818</td>\n",
       "      <td>1116081</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1081823</td>\n",
       "      <td>1116098</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1173884</td>\n",
       "      <td>1197935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162571.13\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1173903</td>\n",
       "      <td>1197935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162571.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1173906</td>\n",
       "      <td>1197933</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162571.13\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0        1        2           3    4  5        6         7         8   \\\n",
       "0  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "1  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "2  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "3  chr1   788438   790667  Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "4  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "5  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "6  chr1  1112984  1113962  Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "7  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "8  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "9  chr1  1194302  1195375   Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "\n",
       "    9     10       11          12       13       14 15 16 17  \\\n",
       "0  635  chr1   HAVANA        gene   586071   827796  .  -  .   \n",
       "1  635  chr1   HAVANA  transcript   594308   827796  .  -  .   \n",
       "2  635  chr1   HAVANA        gene   778770   810060  .  +  .   \n",
       "3  635  chr1   HAVANA  transcript   779056   804875  .  +  .   \n",
       "4  471  chr1   HAVANA        gene  1081818  1116361  .  -  .   \n",
       "5  471  chr1   HAVANA  transcript  1081818  1116081  .  -  .   \n",
       "6  471  chr1   HAVANA  transcript  1081823  1116098  .  -  .   \n",
       "7  650  chr1   HAVANA        gene  1173884  1197935  .  +  .   \n",
       "8  650  chr1   HAVANA  transcript  1173903  1197935  .  +  .   \n",
       "9  650  chr1  ENSEMBL  transcript  1173906  1197933  .  +  .   \n",
       "\n",
       "                                                  18  \n",
       "0  gene_id \"ENSG00000230021.10\"; gene_type \"trans...  \n",
       "1  gene_id \"ENSG00000230021.10\"; transcript_id \"E...  \n",
       "2  gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...  \n",
       "3  gene_id \"ENSG00000237491.8\"; transcript_id \"EN...  \n",
       "4  gene_id \"ENSG00000131591.17\"; gene_type \"prote...  \n",
       "5  gene_id \"ENSG00000131591.17\"; transcript_id \"E...  \n",
       "6  gene_id \"ENSG00000131591.17\"; transcript_id \"E...  \n",
       "7  gene_id \"ENSG00000162571.13\"; gene_type \"prote...  \n",
       "8  gene_id \"ENSG00000162571.13\"; transcript_id \"E...  \n",
       "9  gene_id \"ENSG00000162571.13\"; transcript_id \"E...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# checking that expected count hasn't been altered\n",
    "\n",
    "display(histone_peaks_K9_df[0:10])\n",
    "display(histone_peaks_K27_df[0:10])\n",
    "display(h3k9me3_pp_annotated_df[0:10])\n",
    "display(h3k27me3_pp_annotated_df[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>594308</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778770</td>\n",
       "      <td>810060</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>779056</td>\n",
       "      <td>804875</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1081818</td>\n",
       "      <td>1116361</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210848</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26618172</td>\n",
       "      <td>26618975</td>\n",
       "      <td>Peak_136047</td>\n",
       "      <td>81</td>\n",
       "      <td>.</td>\n",
       "      <td>3.53416</td>\n",
       "      <td>8.12781</td>\n",
       "      <td>6.56647</td>\n",
       "      <td>650</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210849</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26622758</td>\n",
       "      <td>26623232</td>\n",
       "      <td>Peak_148094</td>\n",
       "      <td>75</td>\n",
       "      <td>.</td>\n",
       "      <td>3.39280</td>\n",
       "      <td>7.50822</td>\n",
       "      <td>5.98529</td>\n",
       "      <td>162</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210850</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26622758</td>\n",
       "      <td>26623232</td>\n",
       "      <td>Peak_148094</td>\n",
       "      <td>75</td>\n",
       "      <td>.</td>\n",
       "      <td>3.39280</td>\n",
       "      <td>7.50822</td>\n",
       "      <td>5.98529</td>\n",
       "      <td>162</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210851</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26629403</td>\n",
       "      <td>26630135</td>\n",
       "      <td>Peak_211236</td>\n",
       "      <td>52</td>\n",
       "      <td>.</td>\n",
       "      <td>2.82733</td>\n",
       "      <td>5.20947</td>\n",
       "      <td>3.84296</td>\n",
       "      <td>493</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210852</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26629403</td>\n",
       "      <td>26630135</td>\n",
       "      <td>Peak_211236</td>\n",
       "      <td>52</td>\n",
       "      <td>.</td>\n",
       "      <td>2.82733</td>\n",
       "      <td>5.20947</td>\n",
       "      <td>3.84296</td>\n",
       "      <td>493</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210853 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2            3    4  5        6         7   \\\n",
       "0       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   \n",
       "1       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   \n",
       "2       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   \n",
       "3       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   \n",
       "4       chr1   1112984   1113962   Peak_13060  297  .  7.35106  29.75177   \n",
       "...      ...       ...       ...          ...  ... ..      ...       ...   \n",
       "210848  chrY  26618172  26618975  Peak_136047   81  .  3.53416   8.12781   \n",
       "210849  chrY  26622758  26623232  Peak_148094   75  .  3.39280   7.50822   \n",
       "210850  chrY  26622758  26623232  Peak_148094   75  .  3.39280   7.50822   \n",
       "210851  chrY  26629403  26630135  Peak_211236   52  .  2.82733   5.20947   \n",
       "210852  chrY  26629403  26630135  Peak_211236   52  .  2.82733   5.20947   \n",
       "\n",
       "              8    9     10      11          12        13        14 15 16 17  \\\n",
       "0       18.43085  635  chr1  HAVANA        gene    586071    827796  .  -  .   \n",
       "1       18.43085  635  chr1  HAVANA  transcript    594308    827796  .  -  .   \n",
       "2       18.43085  635  chr1  HAVANA        gene    778770    810060  .  +  .   \n",
       "3       18.43085  635  chr1  HAVANA  transcript    779056    804875  .  +  .   \n",
       "4       27.03791  471  chr1  HAVANA        gene   1081818   1116361  .  -  .   \n",
       "...          ...  ...   ...     ...         ...       ...       ... .. .. ..   \n",
       "210848   6.56647  650  chrY  HAVANA  transcript  26594851  26634652  .  -  .   \n",
       "210849   5.98529  162  chrY  HAVANA        gene  26594851  26634652  .  -  .   \n",
       "210850   5.98529  162  chrY  HAVANA  transcript  26594851  26634652  .  -  .   \n",
       "210851   3.84296  493  chrY  HAVANA        gene  26594851  26634652  .  -  .   \n",
       "210852   3.84296  493  chrY  HAVANA  transcript  26594851  26634652  .  -  .   \n",
       "\n",
       "                                                       18  \n",
       "0       gene_id \"ENSG00000230021.10\"; gene_type \"trans...  \n",
       "1       gene_id \"ENSG00000230021.10\"; transcript_id \"E...  \n",
       "2       gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...  \n",
       "3       gene_id \"ENSG00000237491.8\"; transcript_id \"EN...  \n",
       "4       gene_id \"ENSG00000131591.17\"; gene_type \"prote...  \n",
       "...                                                   ...  \n",
       "210848  gene_id \"ENSG00000237917.1\"; transcript_id \"EN...  \n",
       "210849  gene_id \"ENSG00000237917.1\"; gene_type \"unproc...  \n",
       "210850  gene_id \"ENSG00000237917.1\"; transcript_id \"EN...  \n",
       "210851  gene_id \"ENSG00000237917.1\"; gene_type \"unproc...  \n",
       "210852  gene_id \"ENSG00000237917.1\"; transcript_id \"EN...  \n",
       "\n",
       "[210853 rows x 19 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>594308</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; transcript_id \"E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778770</td>\n",
       "      <td>810060</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>779056</td>\n",
       "      <td>804875</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1081818</td>\n",
       "      <td>1116361</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210848</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26618172</td>\n",
       "      <td>26618975</td>\n",
       "      <td>Peak_136047</td>\n",
       "      <td>81</td>\n",
       "      <td>.</td>\n",
       "      <td>3.53416</td>\n",
       "      <td>8.12781</td>\n",
       "      <td>6.56647</td>\n",
       "      <td>650</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210849</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26622758</td>\n",
       "      <td>26623232</td>\n",
       "      <td>Peak_148094</td>\n",
       "      <td>75</td>\n",
       "      <td>.</td>\n",
       "      <td>3.39280</td>\n",
       "      <td>7.50822</td>\n",
       "      <td>5.98529</td>\n",
       "      <td>162</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210850</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26622758</td>\n",
       "      <td>26623232</td>\n",
       "      <td>Peak_148094</td>\n",
       "      <td>75</td>\n",
       "      <td>.</td>\n",
       "      <td>3.39280</td>\n",
       "      <td>7.50822</td>\n",
       "      <td>5.98529</td>\n",
       "      <td>162</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210851</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26629403</td>\n",
       "      <td>26630135</td>\n",
       "      <td>Peak_211236</td>\n",
       "      <td>52</td>\n",
       "      <td>.</td>\n",
       "      <td>2.82733</td>\n",
       "      <td>5.20947</td>\n",
       "      <td>3.84296</td>\n",
       "      <td>493</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210852</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26629403</td>\n",
       "      <td>26630135</td>\n",
       "      <td>Peak_211236</td>\n",
       "      <td>52</td>\n",
       "      <td>.</td>\n",
       "      <td>2.82733</td>\n",
       "      <td>5.20947</td>\n",
       "      <td>3.84296</td>\n",
       "      <td>493</td>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>26594851</td>\n",
       "      <td>26634652</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237917.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210853 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2            3    4  5        6         7   \\\n",
       "0       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   \n",
       "1       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   \n",
       "2       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   \n",
       "3       chr1    788438    790667   Peak_33943  206  .  4.85493  20.69931   \n",
       "4       chr1   1112984   1113962   Peak_13060  297  .  7.35106  29.75177   \n",
       "...      ...       ...       ...          ...  ... ..      ...       ...   \n",
       "210848  chrY  26618172  26618975  Peak_136047   81  .  3.53416   8.12781   \n",
       "210849  chrY  26622758  26623232  Peak_148094   75  .  3.39280   7.50822   \n",
       "210850  chrY  26622758  26623232  Peak_148094   75  .  3.39280   7.50822   \n",
       "210851  chrY  26629403  26630135  Peak_211236   52  .  2.82733   5.20947   \n",
       "210852  chrY  26629403  26630135  Peak_211236   52  .  2.82733   5.20947   \n",
       "\n",
       "              8    9     10      11          12        13        14 15 16 17  \\\n",
       "0       18.43085  635  chr1  HAVANA        gene    586071    827796  .  -  .   \n",
       "1       18.43085  635  chr1  HAVANA  transcript    594308    827796  .  -  .   \n",
       "2       18.43085  635  chr1  HAVANA        gene    778770    810060  .  +  .   \n",
       "3       18.43085  635  chr1  HAVANA  transcript    779056    804875  .  +  .   \n",
       "4       27.03791  471  chr1  HAVANA        gene   1081818   1116361  .  -  .   \n",
       "...          ...  ...   ...     ...         ...       ...       ... .. .. ..   \n",
       "210848   6.56647  650  chrY  HAVANA  transcript  26594851  26634652  .  -  .   \n",
       "210849   5.98529  162  chrY  HAVANA        gene  26594851  26634652  .  -  .   \n",
       "210850   5.98529  162  chrY  HAVANA  transcript  26594851  26634652  .  -  .   \n",
       "210851   3.84296  493  chrY  HAVANA        gene  26594851  26634652  .  -  .   \n",
       "210852   3.84296  493  chrY  HAVANA  transcript  26594851  26634652  .  -  .   \n",
       "\n",
       "                                                       18  \n",
       "0       gene_id \"ENSG00000230021.10\"; gene_type \"trans...  \n",
       "1       gene_id \"ENSG00000230021.10\"; transcript_id \"E...  \n",
       "2       gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...  \n",
       "3       gene_id \"ENSG00000237491.8\"; transcript_id \"EN...  \n",
       "4       gene_id \"ENSG00000131591.17\"; gene_type \"prote...  \n",
       "...                                                   ...  \n",
       "210848  gene_id \"ENSG00000237917.1\"; transcript_id \"EN...  \n",
       "210849  gene_id \"ENSG00000237917.1\"; gene_type \"unproc...  \n",
       "210850  gene_id \"ENSG00000237917.1\"; transcript_id \"EN...  \n",
       "210851  gene_id \"ENSG00000237917.1\"; gene_type \"unproc...  \n",
       "210852  gene_id \"ENSG00000237917.1\"; transcript_id \"EN...  \n",
       "\n",
       "[210853 rows x 19 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(h3k27me3_pp_annotated_df)\n",
    "display(h3k9me3_pp_annotated_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intervals\n",
    "\n",
    "aggregating to the gene level for WGBS and histone pp data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_id \"ENSG00000230021.10\"; gene_type \"transcribed_processed_pseudogene\"; gene_name \"AL669831.3\"; level 2; havana_gene \"OTTHUMG00000191652.4\";\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     0      1      2   3   4  5      6      7        8   9   ...  13    14  \\\n",
       "0  chr1  14709  14710   1   1  +  14709  14710  255,0,0   1  ...   2  chr1   \n",
       "2  chr1  14715  14716   1   1  +  14715  14716  255,0,0   1  ...   2  chr1   \n",
       "4  chr1  14740  14741   1   1  +  14740  14741  255,0,0   1  ...   2  chr1   \n",
       "6  chr1  14746  14747   1   1  +  14746  14747  255,0,0   1  ...   2  chr1   \n",
       "8  chr1  14772  14773   1   1  +  14772  14773  255,0,0   1  ...   2  chr1   \n",
       "\n",
       "       15    16     17     18 19  20  21  \\\n",
       "0  HAVANA  gene  14404  29570  .   -   .   \n",
       "2  HAVANA  gene  14404  29570  .   -   .   \n",
       "4  HAVANA  gene  14404  29570  .   -   .   \n",
       "6  HAVANA  gene  14404  29570  .   -   .   \n",
       "8  HAVANA  gene  14404  29570  .   -   .   \n",
       "\n",
       "                                                  22  \n",
       "0  gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "2  gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "4  gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "6  gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "8  gene_id \"ENSG00000227232.5\"; gene_type \"unproc...  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778770</td>\n",
       "      <td>810060</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1081818</td>\n",
       "      <td>1116361</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1173884</td>\n",
       "      <td>1197935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162571.13\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1770030</td>\n",
       "      <td>1770624</td>\n",
       "      <td>Peak_107498</td>\n",
       "      <td>98</td>\n",
       "      <td>.</td>\n",
       "      <td>3.79372</td>\n",
       "      <td>9.87359</td>\n",
       "      <td>8.21870</td>\n",
       "      <td>315</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1751232</td>\n",
       "      <td>1780457</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000008130.15\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1        2            3    4  5        6         7         8   \\\n",
       "0   chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "2   chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "4   chr1  1112984  1113962   Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "7   chr1  1194302  1195375    Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "10  chr1  1770030  1770624  Peak_107498   98  .  3.79372   9.87359   8.21870   \n",
       "\n",
       "     9     10      11    12       13       14 15 16 17  \\\n",
       "0   635  chr1  HAVANA  gene   586071   827796  .  -  .   \n",
       "2   635  chr1  HAVANA  gene   778770   810060  .  +  .   \n",
       "4   471  chr1  HAVANA  gene  1081818  1116361  .  -  .   \n",
       "7   650  chr1  HAVANA  gene  1173884  1197935  .  +  .   \n",
       "10  315  chr1  HAVANA  gene  1751232  1780457  .  -  .   \n",
       "\n",
       "                                                   18  \n",
       "0   gene_id \"ENSG00000230021.10\"; gene_type \"trans...  \n",
       "2   gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...  \n",
       "4   gene_id \"ENSG00000131591.17\"; gene_type \"prote...  \n",
       "7   gene_id \"ENSG00000162571.13\"; gene_type \"prote...  \n",
       "10  gene_id \"ENSG00000008130.15\"; gene_type \"prote...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230021.10\"; gene_type \"trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>778770</td>\n",
       "      <td>810060</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1081818</td>\n",
       "      <td>1116361</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000131591.17\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1173884</td>\n",
       "      <td>1197935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000162571.13\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1770030</td>\n",
       "      <td>1770624</td>\n",
       "      <td>Peak_107498</td>\n",
       "      <td>98</td>\n",
       "      <td>.</td>\n",
       "      <td>3.79372</td>\n",
       "      <td>9.87359</td>\n",
       "      <td>8.21870</td>\n",
       "      <td>315</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1751232</td>\n",
       "      <td>1780457</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000008130.15\"; gene_type \"prote...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1        2            3    4  5        6         7         8   \\\n",
       "0   chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "2   chr1   788438   790667   Peak_33943  206  .  4.85493  20.69931  18.43085   \n",
       "4   chr1  1112984  1113962   Peak_13060  297  .  7.35106  29.75177  27.03791   \n",
       "7   chr1  1194302  1195375    Peak_6971  362  .  8.27128  36.28568  33.31710   \n",
       "10  chr1  1770030  1770624  Peak_107498   98  .  3.79372   9.87359   8.21870   \n",
       "\n",
       "     9     10      11    12       13       14 15 16 17  \\\n",
       "0   635  chr1  HAVANA  gene   586071   827796  .  -  .   \n",
       "2   635  chr1  HAVANA  gene   778770   810060  .  +  .   \n",
       "4   471  chr1  HAVANA  gene  1081818  1116361  .  -  .   \n",
       "7   650  chr1  HAVANA  gene  1173884  1197935  .  +  .   \n",
       "10  315  chr1  HAVANA  gene  1751232  1780457  .  -  .   \n",
       "\n",
       "                                                   18  \n",
       "0   gene_id \"ENSG00000230021.10\"; gene_type \"trans...  \n",
       "2   gene_id \"ENSG00000237491.8\"; gene_type \"lincRN...  \n",
       "4   gene_id \"ENSG00000131591.17\"; gene_type \"prote...  \n",
       "7   gene_id \"ENSG00000162571.13\"; gene_type \"prote...  \n",
       "10  gene_id \"ENSG00000008130.15\"; gene_type \"prote...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#filtering to keep only \"gene\" rows (removing \"transcript\", \"exon\", etc.)\n",
    "wgbs_annotated_gene_df = wgbs_annotated_df[wgbs_annotated_df[16] == 'gene']\n",
    "h3k9me3_pp_annotated_gene_df = h3k9me3_pp_annotated_df[h3k9me3_pp_annotated_df[12] == 'gene']\n",
    "h3k27me3_pp_annotated_gene_df = h3k27me3_pp_annotated_df[h3k27me3_pp_annotated_df[12] == 'gene']\n",
    "\n",
    "print(h3k9me3_pp_annotated_gene_df[18][0])\n",
    "display(wgbs_annotated_gene_df.head())\n",
    "display(h3k9me3_pp_annotated_gene_df.head())\n",
    "display(h3k27me3_pp_annotated_gene_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extracting gene IDs from annotated dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\julia\\AppData\\Local\\Temp\\ipykernel_15684\\3790687905.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  h3k9me3_pp_annotated_gene_df.loc[:, 'GeneID'] = h3k9me3_pp_annotated_gene_df.iloc[:, -1].str.extract(pattern)\n",
      "C:\\Users\\julia\\AppData\\Local\\Temp\\ipykernel_15684\\3790687905.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  h3k27me3_pp_annotated_gene_df.loc[:, 'GeneID'] = h3k27me3_pp_annotated_gene_df.iloc[:, -1].str.extract(pattern)\n"
     ]
    }
   ],
   "source": [
    "# extracting gene ID\n",
    "h3k9me3_pp_annotated_gene_df = h3k9me3_pp_annotated_df[h3k9me3_pp_annotated_df[12] == 'gene']\n",
    "h3k27me3_pp_annotated_gene_df = h3k27me3_pp_annotated_df[h3k27me3_pp_annotated_df[12] == 'gene']\n",
    "pattern = r'gene_id\\s*\"([^\"]+)\"'\n",
    "h3k9me3_pp_annotated_gene_df.loc[:, 'GeneID'] = h3k9me3_pp_annotated_gene_df.iloc[:, -1].str.extract(pattern)\n",
    "h3k27me3_pp_annotated_gene_df.loc[:, 'GeneID'] = h3k27me3_pp_annotated_gene_df.iloc[:, -1].str.extract(pattern)\n",
    "#wgbs_annotated_gene_df.loc[:, 'GeneID'] = wgbs_annotated_gene_df.iloc[:, -1].str.extract(pattern)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Aggregating data by gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "WGBS aggregating by gene. 1st attempt = aggregating methylation percentage\n",
    "\n",
    "\n",
    "do I only aggregate CpG-rich regions? How to approach this? filter by coverage? Ensure reliable results?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>GeneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14709</td>\n",
       "      <td>14710</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.5\"; gene_type \"unproc...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999957</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9722512</td>\n",
       "      <td>9722513</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>9722512</td>\n",
       "      <td>9722513</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>9651732</td>\n",
       "      <td>9729114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000171608.15\"; gene_type \"prote...</td>\n",
       "      <td>ENSG00000171608.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999963</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9722513</td>\n",
       "      <td>9722514</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>9722513</td>\n",
       "      <td>9722514</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>9651732</td>\n",
       "      <td>9729114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000171608.15\"; gene_type \"prote...</td>\n",
       "      <td>ENSG00000171608.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999969</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9722516</td>\n",
       "      <td>9722517</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>9722516</td>\n",
       "      <td>9722517</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>9651732</td>\n",
       "      <td>9729114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000171608.15\"; gene_type \"prote...</td>\n",
       "      <td>ENSG00000171608.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999975</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9722517</td>\n",
       "      <td>9722518</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>9722517</td>\n",
       "      <td>9722518</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>9651732</td>\n",
       "      <td>9729114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000171608.15\"; gene_type \"prote...</td>\n",
       "      <td>ENSG00000171608.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999991</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9722532</td>\n",
       "      <td>9722533</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>9722532</td>\n",
       "      <td>9722533</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>9651732</td>\n",
       "      <td>9729114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000171608.15\"; gene_type \"prote...</td>\n",
       "      <td>ENSG00000171608.15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>254680 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0        1        2  3  4  5        6        7        8  9  ...  \\\n",
       "0       chr1    14709    14710  1  1  +    14709    14710  255,0,0  1  ...   \n",
       "2       chr1    14715    14716  1  1  +    14715    14716  255,0,0  1  ...   \n",
       "4       chr1    14740    14741  1  1  +    14740    14741  255,0,0  1  ...   \n",
       "6       chr1    14746    14747  1  1  +    14746    14747  255,0,0  1  ...   \n",
       "8       chr1    14772    14773  1  1  +    14772    14773  255,0,0  1  ...   \n",
       "...      ...      ...      ... .. .. ..      ...      ...      ... ..  ...   \n",
       "999957  chr1  9722512  9722513  1  1  +  9722512  9722513  255,0,0  1  ...   \n",
       "999963  chr1  9722513  9722514  1  3  -  9722513  9722514  255,0,0  3  ...   \n",
       "999969  chr1  9722516  9722517  1  2  +  9722516  9722517  255,0,0  2  ...   \n",
       "999975  chr1  9722517  9722518  1  3  -  9722517  9722518  255,0,0  3  ...   \n",
       "999991  chr1  9722532  9722533  1  2  +  9722532  9722533  255,0,0  2  ...   \n",
       "\n",
       "          14      15    16       17       18 19 20  21  \\\n",
       "0       chr1  HAVANA  gene    14404    29570  .  -   .   \n",
       "2       chr1  HAVANA  gene    14404    29570  .  -   .   \n",
       "4       chr1  HAVANA  gene    14404    29570  .  -   .   \n",
       "6       chr1  HAVANA  gene    14404    29570  .  -   .   \n",
       "8       chr1  HAVANA  gene    14404    29570  .  -   .   \n",
       "...      ...     ...   ...      ...      ... .. ..  ..   \n",
       "999957  chr1  HAVANA  gene  9651732  9729114  .  +   .   \n",
       "999963  chr1  HAVANA  gene  9651732  9729114  .  +   .   \n",
       "999969  chr1  HAVANA  gene  9651732  9729114  .  +   .   \n",
       "999975  chr1  HAVANA  gene  9651732  9729114  .  +   .   \n",
       "999991  chr1  HAVANA  gene  9651732  9729114  .  +   .   \n",
       "\n",
       "                                                       22              GeneID  \n",
       "0       gene_id \"ENSG00000227232.5\"; gene_type \"unproc...   ENSG00000227232.5  \n",
       "2       gene_id \"ENSG00000227232.5\"; gene_type \"unproc...   ENSG00000227232.5  \n",
       "4       gene_id \"ENSG00000227232.5\"; gene_type \"unproc...   ENSG00000227232.5  \n",
       "6       gene_id \"ENSG00000227232.5\"; gene_type \"unproc...   ENSG00000227232.5  \n",
       "8       gene_id \"ENSG00000227232.5\"; gene_type \"unproc...   ENSG00000227232.5  \n",
       "...                                                   ...                 ...  \n",
       "999957  gene_id \"ENSG00000171608.15\"; gene_type \"prote...  ENSG00000171608.15  \n",
       "999963  gene_id \"ENSG00000171608.15\"; gene_type \"prote...  ENSG00000171608.15  \n",
       "999969  gene_id \"ENSG00000171608.15\"; gene_type \"prote...  ENSG00000171608.15  \n",
       "999975  gene_id \"ENSG00000171608.15\"; gene_type \"prote...  ENSG00000171608.15  \n",
       "999991  gene_id \"ENSG00000171608.15\"; gene_type \"prote...  ENSG00000171608.15  \n",
       "\n",
       "[254680 rows x 24 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aggregating WGBS data by gene and average pct methylation per bp in gene (avg pct per gene bp)\n",
    "wgbs_annotated_gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14715</td>\n",
       "      <td>14716</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14740</td>\n",
       "      <td>14741</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14746</td>\n",
       "      <td>14747</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14772</td>\n",
       "      <td>14773</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14774</td>\n",
       "      <td>14775</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14774</td>\n",
       "      <td>14775</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14790</td>\n",
       "      <td>14791</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>14790</td>\n",
       "      <td>14791</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15643</td>\n",
       "      <td>15644</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>15643</td>\n",
       "      <td>15644</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16243</td>\n",
       "      <td>16244</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>16243</td>\n",
       "      <td>16244</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16554</td>\n",
       "      <td>16555</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>16554</td>\n",
       "      <td>16555</td>\n",
       "      <td>255,55,0</td>\n",
       "      <td>7</td>\n",
       "      <td>85</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>10</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0      1      2  3  4  5      6      7         8  9   10  11  12  13  \\\n",
       "2   chr1  14715  14716  1  1  +  14715  14716   255,0,0  1  100  CG  CG   2   \n",
       "4   chr1  14740  14741  1  1  +  14740  14741   255,0,0  1  100  CG  CG   2   \n",
       "6   chr1  14746  14747  1  1  +  14746  14747   255,0,0  1  100  CG  CG   2   \n",
       "8   chr1  14772  14773  1  1  +  14772  14773   255,0,0  1  100  CG  CG   2   \n",
       "10  chr1  14774  14775  1  1  +  14774  14775   255,0,0  1  100  CG  CG   2   \n",
       "12  chr1  14790  14791  1  1  +  14790  14791   255,0,0  1  100  CG  CG   2   \n",
       "14  chr1  15643  15644  1  1  -  15643  15644   255,0,0  1  100  CG  CG   2   \n",
       "16  chr1  16243  16244  1  1  -  16243  16244   255,0,0  1  100  CG  CG   2   \n",
       "18  chr1  16554  16555  1  7  +  16554  16555  255,55,0  7   85  CG  CG  10   \n",
       "\n",
       "      14  \n",
       "2   chr1  \n",
       "4   chr1  \n",
       "6   chr1  \n",
       "8   chr1  \n",
       "10  chr1  \n",
       "12  chr1  \n",
       "14  chr1  \n",
       "16  chr1  \n",
       "18  chr1  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgbs_annotated_gene_df.iloc[1:10,0:15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "## need to fix logic here\n",
    "# get pct and get the average over each bp\n",
    "wgbs_pct_avg_per_gene = wgbs_annotated_gene_df.groupby('GeneID').agg({10: \"mean\"}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000007923.15</td>\n",
       "      <td>90.242105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000008128.22</td>\n",
       "      <td>83.959079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000008130.15</td>\n",
       "      <td>75.072961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000041988.15</td>\n",
       "      <td>71.348031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000049239.12</td>\n",
       "      <td>80.770075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>ENSG00000284745.1</td>\n",
       "      <td>28.580292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>ENSG00000284747.1</td>\n",
       "      <td>92.269231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>ENSG00000285629.1</td>\n",
       "      <td>46.465854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>ENSG00000285812.1</td>\n",
       "      <td>93.048077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>ENSG00000285945.1</td>\n",
       "      <td>16.409091</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>307 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GeneID         10\n",
       "0    ENSG00000007923.15  90.242105\n",
       "1    ENSG00000008128.22  83.959079\n",
       "2    ENSG00000008130.15  75.072961\n",
       "3    ENSG00000041988.15  71.348031\n",
       "4    ENSG00000049239.12  80.770075\n",
       "..                  ...        ...\n",
       "302   ENSG00000284745.1  28.580292\n",
       "303   ENSG00000284747.1  92.269231\n",
       "304   ENSG00000285629.1  46.465854\n",
       "305   ENSG00000285812.1  93.048077\n",
       "306   ENSG00000285945.1  16.409091\n",
       "\n",
       "[307 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "count    307.000000\n",
       "mean      57.631266\n",
       "std       28.367165\n",
       "min        0.000000\n",
       "25%       35.248829\n",
       "50%       59.333333\n",
       "75%       83.562500\n",
       "max      100.000000\n",
       "Name: 10, dtype: float64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display(wgbs_pct_avg_per_gene)\n",
    "wgbs_pct_avg_per_gene[10].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histone peak aggregation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>PeakCount</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000971.15</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000001084.12</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000001629.9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8451</th>\n",
       "      <td>ENSG00000285978.1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8452</th>\n",
       "      <td>ENSG00000285981.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8453</th>\n",
       "      <td>ENSG00000285986.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8454</th>\n",
       "      <td>ENSG00000285987.1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8455</th>\n",
       "      <td>ENSG00000285989.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8456 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GeneID  PeakCount\n",
       "0     ENSG00000000457.13          1\n",
       "1     ENSG00000000460.16          1\n",
       "2     ENSG00000000971.15          6\n",
       "3     ENSG00000001084.12          5\n",
       "4      ENSG00000001629.9          3\n",
       "...                  ...        ...\n",
       "8451   ENSG00000285978.1          7\n",
       "8452   ENSG00000285981.1          2\n",
       "8453   ENSG00000285986.1          1\n",
       "8454   ENSG00000285987.1          5\n",
       "8455   ENSG00000285989.1          1\n",
       "\n",
       "[8456 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>PeakCount</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000971.15</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000001084.12</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000001629.9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8451</th>\n",
       "      <td>ENSG00000285978.1</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8452</th>\n",
       "      <td>ENSG00000285981.1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8453</th>\n",
       "      <td>ENSG00000285986.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8454</th>\n",
       "      <td>ENSG00000285987.1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8455</th>\n",
       "      <td>ENSG00000285989.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8456 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GeneID  PeakCount\n",
       "0     ENSG00000000457.13          1\n",
       "1     ENSG00000000460.16          1\n",
       "2     ENSG00000000971.15          6\n",
       "3     ENSG00000001084.12          5\n",
       "4      ENSG00000001629.9          3\n",
       "...                  ...        ...\n",
       "8451   ENSG00000285978.1          7\n",
       "8452   ENSG00000285981.1          2\n",
       "8453   ENSG00000285986.1          1\n",
       "8454   ENSG00000285987.1          5\n",
       "8455   ENSG00000285989.1          1\n",
       "\n",
       "[8456 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Histone aggregating by gene\n",
    "# sum peaks per gene? \n",
    "# what do the other columns represent?\n",
    "\n",
    "k9_peaks_per_gene = h3k9me3_pp_annotated_gene_df.groupby('GeneID').size().reset_index(name='PeakCount')\n",
    "k27_peaks_per_gene = h3k27me3_pp_annotated_gene_df.groupby('GeneID').size().reset_index(name='PeakCount')\n",
    "\n",
    "display(k9_peaks_per_gene)\n",
    "display(k27_peaks_per_gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA aggregation\n",
    "\n",
    "RNA quant file contains ensemble genes\n",
    "Annotate ensemble genes with corresponding V29 gene\n",
    "A mapping file has been created to extract the chromosome, start site, end site, enseml name and V29 name\n",
    "They will be merged below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# loading the mapping file and gene quant file\n",
    "rna_quant_annotated = pd.read_csv('HepG2_data/HepG2_DNAm/annotated/rna_gene_quant_with_annotations.csv', header=[0])\n",
    "rna_quant_annotated_genes = rna_quant_annotated[rna_quant_annotated['gene_name'].notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>...</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>ENST00000373020.8,ENST00000494424.1,ENST000004...</td>\n",
       "      <td>2179.14</td>\n",
       "      <td>2007.92</td>\n",
       "      <td>2019.00</td>\n",
       "      <td>36.23</td>\n",
       "      <td>68.31</td>\n",
       "      <td>2019.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>34.91</td>\n",
       "      <td>...</td>\n",
       "      <td>3.334520e+01</td>\n",
       "      <td>36.533100</td>\n",
       "      <td>0.015839</td>\n",
       "      <td>6.478720e+01</td>\n",
       "      <td>70.974500</td>\n",
       "      <td>0.015831</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100627109.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>ENST00000373031.4,ENST00000485971.1</td>\n",
       "      <td>940.50</td>\n",
       "      <td>769.31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.12</td>\n",
       "      <td>...</td>\n",
       "      <td>1.760980e-03</td>\n",
       "      <td>0.313410</td>\n",
       "      <td>0.511119</td>\n",
       "      <td>3.288750e-03</td>\n",
       "      <td>0.609651</td>\n",
       "      <td>0.511279</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100584802.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>TNMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>ENST00000371582.8,ENST00000371584.8,ENST000003...</td>\n",
       "      <td>1078.03</td>\n",
       "      <td>906.82</td>\n",
       "      <td>887.00</td>\n",
       "      <td>35.24</td>\n",
       "      <td>66.45</td>\n",
       "      <td>887.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>34.27</td>\n",
       "      <td>...</td>\n",
       "      <td>3.199170e+01</td>\n",
       "      <td>36.595700</td>\n",
       "      <td>0.023183</td>\n",
       "      <td>6.222900e+01</td>\n",
       "      <td>71.169600</td>\n",
       "      <td>0.023156</td>\n",
       "      <td>chr20</td>\n",
       "      <td>50934867.0</td>\n",
       "      <td>50958555.0</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>ENST00000367770.5,ENST00000367771.10,ENST00000...</td>\n",
       "      <td>3488.94</td>\n",
       "      <td>3317.73</td>\n",
       "      <td>258.00</td>\n",
       "      <td>2.80</td>\n",
       "      <td>5.28</td>\n",
       "      <td>258.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.79</td>\n",
       "      <td>...</td>\n",
       "      <td>2.313660e+00</td>\n",
       "      <td>3.287400</td>\n",
       "      <td>0.060300</td>\n",
       "      <td>4.505810e+00</td>\n",
       "      <td>6.399090</td>\n",
       "      <td>0.060332</td>\n",
       "      <td>chr1</td>\n",
       "      <td>169849631.0</td>\n",
       "      <td>169894267.0</td>\n",
       "      <td>SCYL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>ENST00000286031.10,ENST00000359326.8,ENST00000...</td>\n",
       "      <td>2911.75</td>\n",
       "      <td>2740.54</td>\n",
       "      <td>879.00</td>\n",
       "      <td>11.56</td>\n",
       "      <td>21.79</td>\n",
       "      <td>879.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.50</td>\n",
       "      <td>...</td>\n",
       "      <td>1.024230e+01</td>\n",
       "      <td>12.846600</td>\n",
       "      <td>0.039000</td>\n",
       "      <td>1.991680e+01</td>\n",
       "      <td>24.984000</td>\n",
       "      <td>0.038972</td>\n",
       "      <td>chr1</td>\n",
       "      <td>169662007.0</td>\n",
       "      <td>169854080.0</td>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59424</th>\n",
       "      <td>ENSG00000285990.1</td>\n",
       "      <td>ENST00000649331.1</td>\n",
       "      <td>647.00</td>\n",
       "      <td>475.81</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.11</td>\n",
       "      <td>1.80</td>\n",
       "      <td>1.60</td>\n",
       "      <td>0.20</td>\n",
       "      <td>...</td>\n",
       "      <td>5.294150e-06</td>\n",
       "      <td>0.530734</td>\n",
       "      <td>0.602857</td>\n",
       "      <td>1.030280e-05</td>\n",
       "      <td>1.032140</td>\n",
       "      <td>0.602870</td>\n",
       "      <td>chr14</td>\n",
       "      <td>19244904.0</td>\n",
       "      <td>19269380.0</td>\n",
       "      <td>AL589743.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59425</th>\n",
       "      <td>ENSG00000285991.1</td>\n",
       "      <td>ENST00000647612.1</td>\n",
       "      <td>5065.00</td>\n",
       "      <td>4893.78</td>\n",
       "      <td>14.93</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.21</td>\n",
       "      <td>17.18</td>\n",
       "      <td>5.91</td>\n",
       "      <td>0.13</td>\n",
       "      <td>...</td>\n",
       "      <td>4.175390e-02</td>\n",
       "      <td>0.234562</td>\n",
       "      <td>0.266380</td>\n",
       "      <td>7.841970e-02</td>\n",
       "      <td>0.453671</td>\n",
       "      <td>0.266300</td>\n",
       "      <td>chr6</td>\n",
       "      <td>149817937.0</td>\n",
       "      <td>149896011.0</td>\n",
       "      <td>AL355312.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59426</th>\n",
       "      <td>ENSG00000285992.1</td>\n",
       "      <td>ENST00000648949.1</td>\n",
       "      <td>956.00</td>\n",
       "      <td>784.78</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>...</td>\n",
       "      <td>1.021180e-06</td>\n",
       "      <td>0.131430</td>\n",
       "      <td>0.654743</td>\n",
       "      <td>2.449570e-06</td>\n",
       "      <td>0.255766</td>\n",
       "      <td>0.654939</td>\n",
       "      <td>chr8</td>\n",
       "      <td>47129262.0</td>\n",
       "      <td>47132628.0</td>\n",
       "      <td>AC120036.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59427</th>\n",
       "      <td>ENSG00000285993.1</td>\n",
       "      <td>ENST00000650266.1</td>\n",
       "      <td>1246.00</td>\n",
       "      <td>1074.78</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000310e-06</td>\n",
       "      <td>0.095424</td>\n",
       "      <td>0.654978</td>\n",
       "      <td>3.892870e-06</td>\n",
       "      <td>0.185655</td>\n",
       "      <td>0.655171</td>\n",
       "      <td>chr18</td>\n",
       "      <td>46409197.0</td>\n",
       "      <td>46410645.0</td>\n",
       "      <td>AC018931.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59428</th>\n",
       "      <td>ENSG00000285994.1</td>\n",
       "      <td>ENST00000648650.1</td>\n",
       "      <td>3732.00</td>\n",
       "      <td>3560.78</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>...</td>\n",
       "      <td>1.194670e-07</td>\n",
       "      <td>0.028976</td>\n",
       "      <td>0.655903</td>\n",
       "      <td>2.323160e-07</td>\n",
       "      <td>0.056364</td>\n",
       "      <td>0.655975</td>\n",
       "      <td>chr10</td>\n",
       "      <td>12563151.0</td>\n",
       "      <td>12567351.0</td>\n",
       "      <td>AL731559.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58721 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id                                   transcript_id(s)  \\\n",
       "649    ENSG00000000003.14  ENST00000373020.8,ENST00000494424.1,ENST000004...   \n",
       "650     ENSG00000000005.5                ENST00000373031.4,ENST00000485971.1   \n",
       "651    ENSG00000000419.12  ENST00000371582.8,ENST00000371584.8,ENST000003...   \n",
       "652    ENSG00000000457.13  ENST00000367770.5,ENST00000367771.10,ENST00000...   \n",
       "653    ENSG00000000460.16  ENST00000286031.10,ENST00000359326.8,ENST00000...   \n",
       "...                   ...                                                ...   \n",
       "59424   ENSG00000285990.1                                  ENST00000649331.1   \n",
       "59425   ENSG00000285991.1                                  ENST00000647612.1   \n",
       "59426   ENSG00000285992.1                                  ENST00000648949.1   \n",
       "59427   ENSG00000285993.1                                  ENST00000650266.1   \n",
       "59428   ENSG00000285994.1                                  ENST00000648650.1   \n",
       "\n",
       "        length  effective_length  expected_count    TPM   FPKM  \\\n",
       "649    2179.14           2007.92         2019.00  36.23  68.31   \n",
       "650     940.50            769.31            0.00   0.00   0.00   \n",
       "651    1078.03            906.82          887.00  35.24  66.45   \n",
       "652    3488.94           3317.73          258.00   2.80   5.28   \n",
       "653    2911.75           2740.54          879.00  11.56  21.79   \n",
       "...        ...               ...             ...    ...    ...   \n",
       "59424   647.00            475.81            0.75   0.06   0.11   \n",
       "59425  5065.00           4893.78           14.93   0.11   0.21   \n",
       "59426   956.00            784.78            0.00   0.00   0.00   \n",
       "59427  1246.00           1074.78            0.00   0.00   0.00   \n",
       "59428  3732.00           3560.78            0.00   0.00   0.00   \n",
       "\n",
       "       posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \\\n",
       "649                 2019.00                                   0.00    34.91   \n",
       "650                    0.00                                   0.00     0.12   \n",
       "651                  887.00                                   0.00    34.27   \n",
       "652                  258.00                                   0.00     2.79   \n",
       "653                  879.00                                   0.00    11.50   \n",
       "...                     ...                                    ...      ...   \n",
       "59424                  1.80                                   1.60     0.20   \n",
       "59425                 17.18                                   5.91     0.13   \n",
       "59426                  0.00                                   0.00     0.04   \n",
       "59427                  0.00                                   0.00     0.03   \n",
       "59428                  0.00                                   0.00     0.01   \n",
       "\n",
       "       ...  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "649    ...        3.334520e+01           36.533100   \n",
       "650    ...        1.760980e-03            0.313410   \n",
       "651    ...        3.199170e+01           36.595700   \n",
       "652    ...        2.313660e+00            3.287400   \n",
       "653    ...        1.024230e+01           12.846600   \n",
       "...    ...                 ...                 ...   \n",
       "59424  ...        5.294150e-06            0.530734   \n",
       "59425  ...        4.175390e-02            0.234562   \n",
       "59426  ...        1.021180e-06            0.131430   \n",
       "59427  ...        2.000310e-06            0.095424   \n",
       "59428  ...        1.194670e-07            0.028976   \n",
       "\n",
       "       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "649                                 0.015839         6.478720e+01   \n",
       "650                                 0.511119         3.288750e-03   \n",
       "651                                 0.023183         6.222900e+01   \n",
       "652                                 0.060300         4.505810e+00   \n",
       "653                                 0.039000         1.991680e+01   \n",
       "...                                      ...                  ...   \n",
       "59424                               0.602857         1.030280e-05   \n",
       "59425                               0.266380         7.841970e-02   \n",
       "59426                               0.654743         2.449570e-06   \n",
       "59427                               0.654978         3.892870e-06   \n",
       "59428                               0.655903         2.323160e-07   \n",
       "\n",
       "       FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \\\n",
       "649              70.974500                                0.015831   \n",
       "650               0.609651                                0.511279   \n",
       "651              71.169600                                0.023156   \n",
       "652               6.399090                                0.060332   \n",
       "653              24.984000                                0.038972   \n",
       "...                    ...                                     ...   \n",
       "59424             1.032140                                0.602870   \n",
       "59425             0.453671                                0.266300   \n",
       "59426             0.255766                                0.654939   \n",
       "59427             0.185655                                0.655171   \n",
       "59428             0.056364                                0.655975   \n",
       "\n",
       "       chromosome        start          end   gene_name  \n",
       "649          chrX  100627109.0  100639991.0      TSPAN6  \n",
       "650          chrX  100584802.0  100599885.0        TNMD  \n",
       "651         chr20   50934867.0   50958555.0        DPM1  \n",
       "652          chr1  169849631.0  169894267.0       SCYL3  \n",
       "653          chr1  169662007.0  169854080.0    C1orf112  \n",
       "...           ...          ...          ...         ...  \n",
       "59424       chr14   19244904.0   19269380.0  AL589743.7  \n",
       "59425        chr6  149817937.0  149896011.0  AL355312.5  \n",
       "59426        chr8   47129262.0   47132628.0  AC120036.5  \n",
       "59427       chr18   46409197.0   46410645.0  AC018931.1  \n",
       "59428       chr10   12563151.0   12567351.0  AL731559.1  \n",
       "\n",
       "[58721 rows x 21 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_quant_annotated_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA is now annotated. We can group by gene and aggregate by expected count."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_count_per_gene = rna_quant_annotated_genes.groupby('gene_id').agg({'expected_count': \"sum\"}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>expected_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.14</td>\n",
       "      <td>2019.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.5</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "      <td>887.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>258.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>879.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58716</th>\n",
       "      <td>ENSG00000285990.1</td>\n",
       "      <td>0.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58717</th>\n",
       "      <td>ENSG00000285991.1</td>\n",
       "      <td>14.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58718</th>\n",
       "      <td>ENSG00000285992.1</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58719</th>\n",
       "      <td>ENSG00000285993.1</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58720</th>\n",
       "      <td>ENSG00000285994.1</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58721 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id  expected_count\n",
       "0      ENSG00000000003.14         2019.00\n",
       "1       ENSG00000000005.5            0.00\n",
       "2      ENSG00000000419.12          887.00\n",
       "3      ENSG00000000457.13          258.00\n",
       "4      ENSG00000000460.16          879.00\n",
       "...                   ...             ...\n",
       "58716   ENSG00000285990.1            0.75\n",
       "58717   ENSG00000285991.1           14.93\n",
       "58718   ENSG00000285992.1            0.00\n",
       "58719   ENSG00000285993.1            0.00\n",
       "58720   ENSG00000285994.1            0.00\n",
       "\n",
       "[58721 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count     58721.000000\n",
      "mean        245.880392\n",
      "std        3587.987114\n",
      "min           0.000000\n",
      "25%           0.000000\n",
      "50%           0.000000\n",
      "75%          11.000000\n",
      "max      532366.000000\n",
      "Name: expected_count, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "display(rna_count_per_gene)\n",
    "print(rna_count_per_gene['expected_count'].describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalisation\n",
    "\n",
    "Normalisation will utilise 'light-house genes'... XXX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For .bigBed files\n",
    "#bb = pybedtools.BedTool('HepG2_DNAm/ENCFF549SNG.bigBed')\n",
    "#for feature in bb:\n",
    "    # process feature, which is an interval in the BED file\n",
    "    #print(feature)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "honours",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
