{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Visualisation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook aims to explore the data retrived from ENCODE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  File accession   File format File type File format type  \\\n",
      "0    ENCFF690FNR     bed bed9+       bed            bed9+   \n",
      "1    ENCFF884UAE        bigWig    bigWig              NaN   \n",
      "2    ENCFF464CWC        bigWig    bigWig              NaN   \n",
      "3    ENCFF549SNG  bigBed bed9+    bigBed            bed9+   \n",
      "4    ENCFF840XVU     bed bed9+       bed            bed9+   \n",
      "\n",
      "                             Output type File assembly Experiment accession  \\\n",
      "0               methylation state at CpG        GRCh38          ENCSR786DCL   \n",
      "1   plus strand methylation state at CpG        GRCh38          ENCSR786DCL   \n",
      "2  minus strand methylation state at CpG        GRCh38          ENCSR786DCL   \n",
      "3               methylation state at CpG        GRCh38          ENCSR786DCL   \n",
      "4               methylation state at CpG        GRCh38          ENCSR625HZA   \n",
      "\n",
      "  Assay                    Donor(s) Biosample term id  ... Platform  \\\n",
      "0  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   \n",
      "1  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   \n",
      "2  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   \n",
      "3  WGBS  /human-donors/ENCDO000AAC/       EFO:0001187  ...      NaN   \n",
      "4  WGBS  /human-donors/ENCDO336AAA/       EFO:0005723  ...      NaN   \n",
      "\n",
      "  Controlled by File Status  \\\n",
      "0           NaN    released   \n",
      "1           NaN    released   \n",
      "2           NaN    released   \n",
      "3           NaN    released   \n",
      "4           NaN    released   \n",
      "\n",
      "                                              s3_uri  \\\n",
      "0  s3://encode-public/2021/07/17/dccc66ef-985f-4e...   \n",
      "1  s3://encode-public/2021/07/17/8407f321-a9f9-4d...   \n",
      "2  s3://encode-public/2021/07/17/b72fba8b-4ccf-47...   \n",
      "3  s3://encode-public/2021/07/17/6aff99be-3685-4a...   \n",
      "4  s3://encode-public/2021/07/20/bfc10eed-ce81-48...   \n",
      "\n",
      "                                           Azure URL    File analysis title  \\\n",
      "0  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "1  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "2  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "3  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "4  https://datasetencode.blob.core.windows.net/da...  ENCODE4 v1.1.6 GRCh38   \n",
      "\n",
      "   File analysis status                 Audit WARNING  Audit NOT_COMPLIANT  \\\n",
      "0              released  low lambda C conversion rate                  NaN   \n",
      "1              released  low lambda C conversion rate                  NaN   \n",
      "2              released  low lambda C conversion rate                  NaN   \n",
      "3              released  low lambda C conversion rate                  NaN   \n",
      "4              released  low lambda C conversion rate                  NaN   \n",
      "\n",
      "                        Audit ERROR  \n",
      "0                               NaN  \n",
      "1                               NaN  \n",
      "2                               NaN  \n",
      "3                               NaN  \n",
      "4  missing lambda C conversion rate  \n",
      "\n",
      "[5 rows x 59 columns]\n"
     ]
    }
   ],
   "source": [
    "# loading the data\n",
    "import pandas as pd\n",
    "\n",
    "metadata_df = pd.read_csv('HepG2_data/HepG2_DNAm/metadata.tsv', sep='\\t')\n",
    "print(metadata_df.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "WGBS - methylation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading WGBS data (methylation)\n",
    "\n",
    "WGBS_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF690FNR.bed', sep='\\t', names=['refChr', 'startChr', 'endChr', 'name', 'score', 'strand', 'startThick', 'endThick', 'rgb', 'reads', 'pctMe', 'refGen', 'sampGen', 'qualScore'])\n",
    "WGBS_plus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF884UAE.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'pctMe'])\n",
    "WGBS_minus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF464CWC.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'pctMe'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns in big bed files (in FNR bed file):\n",
    "- Reference chromosome or scaffold\n",
    "- Start position in chromosome\n",
    "- End position in chromosome\n",
    "- Name of item\n",
    "- Score from 0-1000. Capped number of reads\n",
    "- Strandedness, plus (+), minus (-), or unknown (?)\n",
    "- Start of where display should be thick\n",
    "- End of where display should be thick\n",
    "- Color value (RGB)\n",
    "- Coverage, or number of reads\n",
    "- Percentage of reads that show methylation at this position in the genome\n",
    "- Reference genotype\n",
    "- Sample genotype\n",
    "- Quality score for genotype call\n",
    "\n",
    "\n",
    "Final column in bigwig files = percent methylation at CpG site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>startChr</th>\n",
       "      <th>endChr</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>startThick</th>\n",
       "      <th>endThick</th>\n",
       "      <th>rgb</th>\n",
       "      <th>reads</th>\n",
       "      <th>pctMe</th>\n",
       "      <th>refGen</th>\n",
       "      <th>sampGen</th>\n",
       "      <th>qualScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10468</td>\n",
       "      <td>10469</td>\n",
       "      <td>0,255,0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CG</td>\n",
       "      <td>MG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>155,255,0</td>\n",
       "      <td>4</td>\n",
       "      <td>25</td>\n",
       "      <td>CG</td>\n",
       "      <td>CK</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10470</td>\n",
       "      <td>10471</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>10483</td>\n",
       "      <td>10484</td>\n",
       "      <td>255,0,0</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>CG</td>\n",
       "      <td>CG</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refChr  startChr  endChr  name  score strand  startThick  endThick  \\\n",
       "0   chr1     10468   10469     1      1      +       10468     10469   \n",
       "1   chr1     10469   10470     1      4      -       10469     10470   \n",
       "2   chr1     10470   10471     1      1      +       10470     10471   \n",
       "3   chr1     10471   10472     1      2      -       10471     10472   \n",
       "4   chr1     10483   10484     1      1      +       10483     10484   \n",
       "\n",
       "         rgb  reads  pctMe refGen sampGen  qualScore  \n",
       "0    0,255,0      1      0     CG      MG          3  \n",
       "1  155,255,0      4     25     CG      CK          5  \n",
       "2    255,0,0      1    100     CG      CG         11  \n",
       "3    255,0,0      2    100     CG      CG          6  \n",
       "4    255,0,0      1    100     CG      CG          4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>pctMe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10009</td>\n",
       "      <td>10012</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10015</td>\n",
       "      <td>10018</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10021</td>\n",
       "      <td>10024</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10027</td>\n",
       "      <td>10030</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10033</td>\n",
       "      <td>10036</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refChr  start    end  pctMe\n",
       "0   chr1  10009  10012    0.0\n",
       "1   chr1  10015  10018    0.0\n",
       "2   chr1  10021  10024    0.0\n",
       "3   chr1  10027  10030    0.0\n",
       "4   chr1  10033  10036    0.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>pctMe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10469</td>\n",
       "      <td>10470</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10471</td>\n",
       "      <td>10472</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10472</td>\n",
       "      <td>10473</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10481</td>\n",
       "      <td>10482</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10484</td>\n",
       "      <td>10485</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refChr  start    end  pctMe\n",
       "0   chr1  10469  10470   25.0\n",
       "1   chr1  10471  10472  100.0\n",
       "2   chr1  10472  10473    0.0\n",
       "3   chr1  10481  10482    0.0\n",
       "4   chr1  10484  10485  100.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(WGBS_df.head())\n",
    "display(WGBS_plus_df.head())\n",
    "display(WGBS_minus_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA data - plus/minus strand and gene quant data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#loading RNA data\n",
    "\n",
    "RNA_geneQuant_df = pd.read_csv('HepG2_data/HepG2_DNAm/ENCFF649XOG.tsv', sep='\\t', header=0)\n",
    "RNA_minus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF194TSC.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'exp'])\n",
    "RNA_plus_df = pd.read_csv('HepG2_data/HepG2_DNAm/processed/ENCFF792BNC.bedGraph', sep='\\t', names=['refChr', 'start', 'end', 'exp'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns for RNA:\n",
    "\n",
    "- chromosome\n",
    "- start\n",
    "- end\n",
    "- gene expression?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190711</td>\n",
       "      <td>190761</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190786</td>\n",
       "      <td>190808</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190808</td>\n",
       "      <td>190836</td>\n",
       "      <td>0.06349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190836</td>\n",
       "      <td>190858</td>\n",
       "      <td>0.04233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>190864</td>\n",
       "      <td>190892</td>\n",
       "      <td>0.04233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198639</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22823877</td>\n",
       "      <td>22823927</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198640</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22824904</td>\n",
       "      <td>22824954</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198641</th>\n",
       "      <td>chrY</td>\n",
       "      <td>22825018</td>\n",
       "      <td>22825067</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198642</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56871389</td>\n",
       "      <td>56871439</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7198643</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56871467</td>\n",
       "      <td>56871517</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7198644 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        refChr     start       end      exp\n",
       "0         chr1    190711    190761  0.02116\n",
       "1         chr1    190786    190808  0.02116\n",
       "2         chr1    190808    190836  0.06349\n",
       "3         chr1    190836    190858  0.04233\n",
       "4         chr1    190864    190892  0.04233\n",
       "...        ...       ...       ...      ...\n",
       "7198639   chrY  22823877  22823927  0.02116\n",
       "7198640   chrY  22824904  22824954  0.02116\n",
       "7198641   chrY  22825018  22825067  0.02116\n",
       "7198642   chrY  56871389  56871439  0.02116\n",
       "7198643   chrY  56871467  56871517  0.02116\n",
       "\n",
       "[7198644 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refChr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14475</td>\n",
       "      <td>14490</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14490</td>\n",
       "      <td>14525</td>\n",
       "      <td>0.04233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14525</td>\n",
       "      <td>14540</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14660</td>\n",
       "      <td>14668</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14668</td>\n",
       "      <td>14675</td>\n",
       "      <td>0.04233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063920</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24213384</td>\n",
       "      <td>24213405</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063921</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24213792</td>\n",
       "      <td>24213821</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063922</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26621781</td>\n",
       "      <td>26621831</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063923</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26621878</td>\n",
       "      <td>26621895</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7063924</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26622511</td>\n",
       "      <td>26622544</td>\n",
       "      <td>0.02116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7063925 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        refChr     start       end      exp\n",
       "0         chr1     14475     14490  0.02116\n",
       "1         chr1     14490     14525  0.04233\n",
       "2         chr1     14525     14540  0.02116\n",
       "3         chr1     14660     14668  0.02116\n",
       "4         chr1     14668     14675  0.04233\n",
       "...        ...       ...       ...      ...\n",
       "7063920   chrY  24213384  24213405  0.02116\n",
       "7063921   chrY  24213792  24213821  0.02116\n",
       "7063922   chrY  26621781  26621831  0.02116\n",
       "7063923   chrY  26621878  26621895  0.02116\n",
       "7063924   chrY  26622511  26622544  0.02116\n",
       "\n",
       "[7063925 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10904</td>\n",
       "      <td>10904</td>\n",
       "      <td>93.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12954</td>\n",
       "      <td>12954</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12956</td>\n",
       "      <td>12956</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12958</td>\n",
       "      <td>12958</td>\n",
       "      <td>82.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12960</td>\n",
       "      <td>12960</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59521</th>\n",
       "      <td>gSpikein_ERCC-00165</td>\n",
       "      <td>tSpikein_ERCC-00165</td>\n",
       "      <td>872.0</td>\n",
       "      <td>700.78</td>\n",
       "      <td>87.0</td>\n",
       "      <td>4.47</td>\n",
       "      <td>8.43</td>\n",
       "      <td>87.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.33</td>\n",
       "      <td>8.42</td>\n",
       "      <td>3.439680</td>\n",
       "      <td>5.244820</td>\n",
       "      <td>0.071704</td>\n",
       "      <td>6.672670</td>\n",
       "      <td>10.182700</td>\n",
       "      <td>0.071661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59522</th>\n",
       "      <td>gSpikein_ERCC-00168</td>\n",
       "      <td>tSpikein_ERCC-00168</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>852.78</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.08</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.002566</td>\n",
       "      <td>0.194005</td>\n",
       "      <td>0.474234</td>\n",
       "      <td>0.005171</td>\n",
       "      <td>0.377749</td>\n",
       "      <td>0.474374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59523</th>\n",
       "      <td>gSpikein_ERCC-00170</td>\n",
       "      <td>tSpikein_ERCC-00170</td>\n",
       "      <td>1023.0</td>\n",
       "      <td>851.78</td>\n",
       "      <td>51.0</td>\n",
       "      <td>2.16</td>\n",
       "      <td>4.07</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.10</td>\n",
       "      <td>4.09</td>\n",
       "      <td>1.548640</td>\n",
       "      <td>2.687830</td>\n",
       "      <td>0.093629</td>\n",
       "      <td>3.011470</td>\n",
       "      <td>5.228390</td>\n",
       "      <td>0.093579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59524</th>\n",
       "      <td>gSpikein_ERCC-00171</td>\n",
       "      <td>tSpikein_ERCC-00171</td>\n",
       "      <td>505.0</td>\n",
       "      <td>333.85</td>\n",
       "      <td>3942.0</td>\n",
       "      <td>425.45</td>\n",
       "      <td>802.16</td>\n",
       "      <td>3942.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>406.87</td>\n",
       "      <td>791.53</td>\n",
       "      <td>394.250000</td>\n",
       "      <td>419.834000</td>\n",
       "      <td>0.010739</td>\n",
       "      <td>767.089000</td>\n",
       "      <td>816.791000</td>\n",
       "      <td>0.010720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59525</th>\n",
       "      <td>gSpikein_phiX174</td>\n",
       "      <td>tSpikein_phiX174</td>\n",
       "      <td>5386.0</td>\n",
       "      <td>5214.78</td>\n",
       "      <td>78337.0</td>\n",
       "      <td>541.27</td>\n",
       "      <td>1020.55</td>\n",
       "      <td>78337.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>517.51</td>\n",
       "      <td>1006.78</td>\n",
       "      <td>513.704000</td>\n",
       "      <td>521.421000</td>\n",
       "      <td>0.002592</td>\n",
       "      <td>999.854000</td>\n",
       "      <td>1013.850000</td>\n",
       "      <td>0.002408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59526 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id     transcript_id(s)  length  effective_length  \\\n",
       "0                    10904                10904    93.0              0.00   \n",
       "1                    12954                12954    94.0              0.00   \n",
       "2                    12956                12956    72.0              0.00   \n",
       "3                    12958                12958    82.0              0.00   \n",
       "4                    12960                12960    73.0              0.00   \n",
       "...                    ...                  ...     ...               ...   \n",
       "59521  gSpikein_ERCC-00165  tSpikein_ERCC-00165   872.0            700.78   \n",
       "59522  gSpikein_ERCC-00168  tSpikein_ERCC-00168  1024.0            852.78   \n",
       "59523  gSpikein_ERCC-00170  tSpikein_ERCC-00170  1023.0            851.78   \n",
       "59524  gSpikein_ERCC-00171  tSpikein_ERCC-00171   505.0            333.85   \n",
       "59525     gSpikein_phiX174     tSpikein_phiX174  5386.0           5214.78   \n",
       "\n",
       "       expected_count     TPM     FPKM  posterior_mean_count  \\\n",
       "0                 0.0    0.00     0.00                   0.0   \n",
       "1                 0.0    0.00     0.00                   0.0   \n",
       "2                 0.0    0.00     0.00                   0.0   \n",
       "3                 0.0    0.00     0.00                   0.0   \n",
       "4                 0.0    0.00     0.00                   0.0   \n",
       "...               ...     ...      ...                   ...   \n",
       "59521            87.0    4.47     8.43                  87.0   \n",
       "59522             1.0    0.04     0.08                   1.0   \n",
       "59523            51.0    2.16     4.07                  51.0   \n",
       "59524          3942.0  425.45   802.16                3942.0   \n",
       "59525         78337.0  541.27  1020.55               78337.0   \n",
       "\n",
       "       posterior_standard_deviation_of_count  pme_TPM  pme_FPKM  \\\n",
       "0                                        0.0     0.00      0.00   \n",
       "1                                        0.0     0.00      0.00   \n",
       "2                                        0.0     0.00      0.00   \n",
       "3                                        0.0     0.00      0.00   \n",
       "4                                        0.0     0.00      0.00   \n",
       "...                                      ...      ...       ...   \n",
       "59521                                    0.0     4.33      8.42   \n",
       "59522                                    0.0     0.08      0.16   \n",
       "59523                                    0.0     2.10      4.09   \n",
       "59524                                    0.0   406.87    791.53   \n",
       "59525                                    0.0   517.51   1006.78   \n",
       "\n",
       "       TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0                0.000000            0.000000   \n",
       "1                0.000000            0.000000   \n",
       "2                0.000000            0.000000   \n",
       "3                0.000000            0.000000   \n",
       "4                0.000000            0.000000   \n",
       "...                   ...                 ...   \n",
       "59521            3.439680            5.244820   \n",
       "59522            0.002566            0.194005   \n",
       "59523            1.548640            2.687830   \n",
       "59524          394.250000          419.834000   \n",
       "59525          513.704000          521.421000   \n",
       "\n",
       "       TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                   0.000000             0.000000   \n",
       "1                                   0.000000             0.000000   \n",
       "2                                   0.000000             0.000000   \n",
       "3                                   0.000000             0.000000   \n",
       "4                                   0.000000             0.000000   \n",
       "...                                      ...                  ...   \n",
       "59521                               0.071704             6.672670   \n",
       "59522                               0.474234             0.005171   \n",
       "59523                               0.093629             3.011470   \n",
       "59524                               0.010739           767.089000   \n",
       "59525                               0.002592           999.854000   \n",
       "\n",
       "       FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \n",
       "0                 0.000000                                0.000000  \n",
       "1                 0.000000                                0.000000  \n",
       "2                 0.000000                                0.000000  \n",
       "3                 0.000000                                0.000000  \n",
       "4                 0.000000                                0.000000  \n",
       "...                    ...                                     ...  \n",
       "59521            10.182700                                0.071661  \n",
       "59522             0.377749                                0.474374  \n",
       "59523             5.228390                                0.093579  \n",
       "59524           816.791000                                0.010720  \n",
       "59525          1013.850000                                0.002408  \n",
       "\n",
       "[59526 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(RNA_plus_df)\n",
    "display(RNA_minus_df)\n",
    "display(RNA_geneQuant_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histone ChIP-Seq data - H3K9me3 and HeK27me3 modifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading histone mod data\n",
    "# H3K9me3\n",
    "histone_pval_K9_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF125NHB.bedGraph', sep='\\t', names=['chrRef', 'start', 'end', 'pval'])\n",
    "histone_peaks_K9_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF170GMN.bed', sep='\\t', header=None)\n",
    "\n",
    "# H3K27me3\n",
    "histone_pval_K27_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF896BFP.bedGraph', sep='\\t', names=['chrRef', 'start', 'end', 'pval'])\n",
    "histone_peaks_K27_df = pd.read_csv('HepG2_data/HepG2_histone/processed/ENCFF841QVP.bed', sep='\\t', header=None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histone columns:\n",
    "- Chromosome \n",
    "- start\n",
    "- end\n",
    "- peak label?\n",
    "- \n",
    "\n",
    "peaks:\n",
    "- chromosome\n",
    "- start\n",
    "- end \n",
    "- replication pct? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>788438</td>\n",
       "      <td>790667</td>\n",
       "      <td>Peak_33943</td>\n",
       "      <td>206</td>\n",
       "      <td>.</td>\n",
       "      <td>4.85493</td>\n",
       "      <td>20.69931</td>\n",
       "      <td>18.43085</td>\n",
       "      <td>635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>991782</td>\n",
       "      <td>992368</td>\n",
       "      <td>Peak_49693</td>\n",
       "      <td>167</td>\n",
       "      <td>.</td>\n",
       "      <td>5.23056</td>\n",
       "      <td>16.71326</td>\n",
       "      <td>14.64974</td>\n",
       "      <td>235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1112984</td>\n",
       "      <td>1113962</td>\n",
       "      <td>Peak_13060</td>\n",
       "      <td>297</td>\n",
       "      <td>.</td>\n",
       "      <td>7.35106</td>\n",
       "      <td>29.75177</td>\n",
       "      <td>27.03791</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1194302</td>\n",
       "      <td>1195375</td>\n",
       "      <td>Peak_6971</td>\n",
       "      <td>362</td>\n",
       "      <td>.</td>\n",
       "      <td>8.27128</td>\n",
       "      <td>36.28568</td>\n",
       "      <td>33.31710</td>\n",
       "      <td>650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1770030</td>\n",
       "      <td>1770624</td>\n",
       "      <td>Peak_107498</td>\n",
       "      <td>98</td>\n",
       "      <td>.</td>\n",
       "      <td>3.79372</td>\n",
       "      <td>9.87359</td>\n",
       "      <td>8.21870</td>\n",
       "      <td>315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148025</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26667979</td>\n",
       "      <td>26668865</td>\n",
       "      <td>Peak_48031</td>\n",
       "      <td>172</td>\n",
       "      <td>.</td>\n",
       "      <td>4.20870</td>\n",
       "      <td>17.27292</td>\n",
       "      <td>15.19134</td>\n",
       "      <td>673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148026</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26668976</td>\n",
       "      <td>26671930</td>\n",
       "      <td>Peak_4</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>6.71736</td>\n",
       "      <td>200.44463</td>\n",
       "      <td>194.12437</td>\n",
       "      <td>2116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148027</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26672410</td>\n",
       "      <td>26672736</td>\n",
       "      <td>Peak_91129</td>\n",
       "      <td>112</td>\n",
       "      <td>.</td>\n",
       "      <td>3.60158</td>\n",
       "      <td>11.25807</td>\n",
       "      <td>9.52160</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148028</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56679663</td>\n",
       "      <td>56680112</td>\n",
       "      <td>Peak_60182</td>\n",
       "      <td>151</td>\n",
       "      <td>.</td>\n",
       "      <td>4.94783</td>\n",
       "      <td>15.15114</td>\n",
       "      <td>13.17521</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148029</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56683855</td>\n",
       "      <td>56684306</td>\n",
       "      <td>Peak_10462</td>\n",
       "      <td>321</td>\n",
       "      <td>.</td>\n",
       "      <td>6.92330</td>\n",
       "      <td>32.11600</td>\n",
       "      <td>29.30741</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148030 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2            3     4  5        6          7  \\\n",
       "0       chr1    788438    790667   Peak_33943   206  .  4.85493   20.69931   \n",
       "1       chr1    991782    992368   Peak_49693   167  .  5.23056   16.71326   \n",
       "2       chr1   1112984   1113962   Peak_13060   297  .  7.35106   29.75177   \n",
       "3       chr1   1194302   1195375    Peak_6971   362  .  8.27128   36.28568   \n",
       "4       chr1   1770030   1770624  Peak_107498    98  .  3.79372    9.87359   \n",
       "...      ...       ...       ...          ...   ... ..      ...        ...   \n",
       "148025  chrY  26667979  26668865   Peak_48031   172  .  4.20870   17.27292   \n",
       "148026  chrY  26668976  26671930       Peak_4  1000  .  6.71736  200.44463   \n",
       "148027  chrY  26672410  26672736   Peak_91129   112  .  3.60158   11.25807   \n",
       "148028  chrY  56679663  56680112   Peak_60182   151  .  4.94783   15.15114   \n",
       "148029  chrY  56683855  56684306   Peak_10462   321  .  6.92330   32.11600   \n",
       "\n",
       "                8     9  \n",
       "0        18.43085   635  \n",
       "1        14.64974   235  \n",
       "2        27.03791   471  \n",
       "3        33.31710   650  \n",
       "4         8.21870   315  \n",
       "...           ...   ...  \n",
       "148025   15.19134   673  \n",
       "148026  194.12437  2116  \n",
       "148027    9.52160    94  \n",
       "148028   13.17521   124  \n",
       "148029   29.30741   227  \n",
       "\n",
       "[148030 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrRef</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>9944</td>\n",
       "      <td>0.00100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9944</td>\n",
       "      <td>9947</td>\n",
       "      <td>0.02630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9947</td>\n",
       "      <td>9950</td>\n",
       "      <td>0.13984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9950</td>\n",
       "      <td>9954</td>\n",
       "      <td>0.39194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9954</td>\n",
       "      <td>10150</td>\n",
       "      <td>0.79457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263757</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>17991</td>\n",
       "      <td>33769</td>\n",
       "      <td>0.00100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263758</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>33769</td>\n",
       "      <td>33984</td>\n",
       "      <td>0.02630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263759</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>33984</td>\n",
       "      <td>34006</td>\n",
       "      <td>0.13984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263760</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>34006</td>\n",
       "      <td>34212</td>\n",
       "      <td>0.02630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251263761</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>34212</td>\n",
       "      <td>34224</td>\n",
       "      <td>0.00100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251263762 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           chrRef  start    end     pval\n",
       "0                            chr1      0   9944  0.00100\n",
       "1                            chr1   9944   9947  0.02630\n",
       "2                            chr1   9947   9950  0.13984\n",
       "3                            chr1   9950   9954  0.39194\n",
       "4                            chr1   9954  10150  0.79457\n",
       "...                           ...    ...    ...      ...\n",
       "251263757  chrY_KI270740v1_random  17991  33769  0.00100\n",
       "251263758  chrY_KI270740v1_random  33769  33984  0.02630\n",
       "251263759  chrY_KI270740v1_random  33984  34006  0.13984\n",
       "251263760  chrY_KI270740v1_random  34006  34212  0.02630\n",
       "251263761  chrY_KI270740v1_random  34212  34224  0.00100\n",
       "\n",
       "[251263762 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100236060</td>\n",
       "      <td>100236175</td>\n",
       "      <td>Peak_91116</td>\n",
       "      <td>33</td>\n",
       "      <td>.</td>\n",
       "      <td>3.38068</td>\n",
       "      <td>3.39518</td>\n",
       "      <td>0.84457</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>100538099</td>\n",
       "      <td>100538408</td>\n",
       "      <td>Peak_9577</td>\n",
       "      <td>65</td>\n",
       "      <td>.</td>\n",
       "      <td>5.03472</td>\n",
       "      <td>6.59658</td>\n",
       "      <td>2.72930</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101135095</td>\n",
       "      <td>101135210</td>\n",
       "      <td>Peak_97001</td>\n",
       "      <td>32</td>\n",
       "      <td>.</td>\n",
       "      <td>3.29575</td>\n",
       "      <td>3.26597</td>\n",
       "      <td>0.83925</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101236811</td>\n",
       "      <td>101237102</td>\n",
       "      <td>Peak_17149</td>\n",
       "      <td>61</td>\n",
       "      <td>.</td>\n",
       "      <td>4.77464</td>\n",
       "      <td>6.16004</td>\n",
       "      <td>2.57973</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>101237336</td>\n",
       "      <td>101237500</td>\n",
       "      <td>Peak_87875</td>\n",
       "      <td>36</td>\n",
       "      <td>.</td>\n",
       "      <td>3.27881</td>\n",
       "      <td>3.62443</td>\n",
       "      <td>1.03439</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51278</th>\n",
       "      <td>chrX</td>\n",
       "      <td>848527</td>\n",
       "      <td>848652</td>\n",
       "      <td>Peak_65447</td>\n",
       "      <td>43</td>\n",
       "      <td>.</td>\n",
       "      <td>3.93203</td>\n",
       "      <td>4.39984</td>\n",
       "      <td>1.48232</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51279</th>\n",
       "      <td>chrX</td>\n",
       "      <td>8742048</td>\n",
       "      <td>8742163</td>\n",
       "      <td>Peak_65448</td>\n",
       "      <td>43</td>\n",
       "      <td>.</td>\n",
       "      <td>3.93203</td>\n",
       "      <td>4.39984</td>\n",
       "      <td>1.48232</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51280</th>\n",
       "      <td>chrX</td>\n",
       "      <td>915821</td>\n",
       "      <td>915961</td>\n",
       "      <td>Peak_5605</td>\n",
       "      <td>77</td>\n",
       "      <td>.</td>\n",
       "      <td>6.13741</td>\n",
       "      <td>7.77377</td>\n",
       "      <td>3.56466</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51281</th>\n",
       "      <td>chrX</td>\n",
       "      <td>9295740</td>\n",
       "      <td>9295902</td>\n",
       "      <td>Peak_31774</td>\n",
       "      <td>54</td>\n",
       "      <td>.</td>\n",
       "      <td>4.48337</td>\n",
       "      <td>5.46977</td>\n",
       "      <td>2.10106</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51282</th>\n",
       "      <td>chrX</td>\n",
       "      <td>966389</td>\n",
       "      <td>966573</td>\n",
       "      <td>Peak_48390</td>\n",
       "      <td>49</td>\n",
       "      <td>.</td>\n",
       "      <td>4.47639</td>\n",
       "      <td>4.98931</td>\n",
       "      <td>1.86485</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51283 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          1          2           3   4  5        6        7  \\\n",
       "0      chr1  100236060  100236175  Peak_91116  33  .  3.38068  3.39518   \n",
       "1      chr1  100538099  100538408   Peak_9577  65  .  5.03472  6.59658   \n",
       "2      chr1  101135095  101135210  Peak_97001  32  .  3.29575  3.26597   \n",
       "3      chr1  101236811  101237102  Peak_17149  61  .  4.77464  6.16004   \n",
       "4      chr1  101237336  101237500  Peak_87875  36  .  3.27881  3.62443   \n",
       "...     ...        ...        ...         ...  .. ..      ...      ...   \n",
       "51278  chrX     848527     848652  Peak_65447  43  .  3.93203  4.39984   \n",
       "51279  chrX    8742048    8742163  Peak_65448  43  .  3.93203  4.39984   \n",
       "51280  chrX     915821     915961   Peak_5605  77  .  6.13741  7.77377   \n",
       "51281  chrX    9295740    9295902  Peak_31774  54  .  4.48337  5.46977   \n",
       "51282  chrX     966389     966573  Peak_48390  49  .  4.47639  4.98931   \n",
       "\n",
       "             8    9  \n",
       "0      0.84457   88  \n",
       "1      2.72930   79  \n",
       "2      0.83925   47  \n",
       "3      2.57973  107  \n",
       "4      1.03439   55  \n",
       "...        ...  ...  \n",
       "51278  1.48232   36  \n",
       "51279  1.48232   26  \n",
       "51280  3.56466   85  \n",
       "51281  2.10106   87  \n",
       "51282  1.86485  111  \n",
       "\n",
       "[51283 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrRef</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>10071</td>\n",
       "      <td>0.30592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10071</td>\n",
       "      <td>10100</td>\n",
       "      <td>0.82513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10100</td>\n",
       "      <td>10127</td>\n",
       "      <td>0.50508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10127</td>\n",
       "      <td>10186</td>\n",
       "      <td>0.18005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10186</td>\n",
       "      <td>10215</td>\n",
       "      <td>0.04769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656086</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>29725</td>\n",
       "      <td>31121</td>\n",
       "      <td>0.30592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656087</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>31121</td>\n",
       "      <td>31236</td>\n",
       "      <td>0.82513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656088</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>31236</td>\n",
       "      <td>36460</td>\n",
       "      <td>0.30592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656089</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>36460</td>\n",
       "      <td>36487</td>\n",
       "      <td>0.82513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59656090</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>36487</td>\n",
       "      <td>36575</td>\n",
       "      <td>0.50508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59656091 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          chrRef  start    end     pval\n",
       "0                           chr1      0  10071  0.30592\n",
       "1                           chr1  10071  10100  0.82513\n",
       "2                           chr1  10100  10127  0.50508\n",
       "3                           chr1  10127  10186  0.18005\n",
       "4                           chr1  10186  10215  0.04769\n",
       "...                          ...    ...    ...      ...\n",
       "59656086  chrY_KI270740v1_random  29725  31121  0.30592\n",
       "59656087  chrY_KI270740v1_random  31121  31236  0.82513\n",
       "59656088  chrY_KI270740v1_random  31236  36460  0.30592\n",
       "59656089  chrY_KI270740v1_random  36460  36487  0.82513\n",
       "59656090  chrY_KI270740v1_random  36487  36575  0.50508\n",
       "\n",
       "[59656091 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(histone_peaks_K9_df)\n",
    "display(histone_pval_K9_df)\n",
    "\n",
    "display(histone_peaks_K27_df)\n",
    "display(histone_pval_K27_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualising\n",
    "\n",
    "To begin, we utilised the HepG2 WGBS methylation state at CPG, RNA-Seq Gene Quantifications, and Histone ChIP-Seq pseudoreplicated peak files (loading them in their original forms). \n",
    "- ENCFF690FNR (first 9 columns only): WGBS\n",
    "- ENCFF194TSC: RNA minus\n",
    "- ENCFF792BNC: RNA plus\n",
    "- ENCFF170GMN: Hist H3K9me3 pseudorep\n",
    "- ENCFF505LOU: Hist H3K27me3 pseudorep"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "WGBS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combining RNA gene quantification file with V29 gene annotations\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Load GTF file - adjust columns names based on your GTF structure\n",
    "gtf_columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "gtf_df = pd.read_csv('yourfile.gtf', sep='\\t', comment='#', names=gtf_columns, usecols=['seqname', 'start', 'end', 'attribute'])\n",
    "\n",
    "# Extract gene IDs and their positions\n",
    "def extract_gene_id(attributes):\n",
    "    return attributes.split('gene_id \"')[1].split('\";')[0]\n",
    "\n",
    "gtf_df['gene_id'] = gtf_df['attribute'].apply(extract_gene_id)\n",
    "gtf_df = gtf_df[['seqname', 'start', 'end', 'gene_id']]\n",
    "\n",
    "# Assuming you have a DataFrame `quant_df` for your quantification data\n",
    "quant_df = pd.read_csv('your_quant_data.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histone ChIP-Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intervals\n",
    "First, annotate data with genes and aggregate at the gene level.\n",
    "\n",
    "Compute the sum of modifications for an entire gene plus its transcription start site. An example model might have the signature: (sum DNA methylation, sum H3K9me3, ...etc) -> (gene expression level). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene annotations for WBGS data - can be applied directly to it using chr, start and end\n",
    "# Gene annotations for RNA data - not sure how to approach this\n",
    "# Gene annotations for hist CS data - do I used pseudo or p-val? annotate both using bedtools intersect"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalisation\n",
    "\n",
    "Normalisation will utilise 'light-house genes'... XXX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For .bigBed files\n",
    "#bb = pybedtools.BedTool('HepG2_DNAm/ENCFF549SNG.bigBed')\n",
    "#for feature in bb:\n",
    "    # process feature, which is an interval in the BED file\n",
    "    #print(feature)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "honours",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
