{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_transformations(y, y_pred, transformation_name):\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    plt.scatter(y, y_pred, alpha=0.5)\n",
    "    plt.plot([y.min(), y.max()], [y.min(), y.max()], \"--r\")\n",
    "    plt.xlabel(\"Actual Counts\")\n",
    "    plt.ylabel(\"Predicted Counts\")\n",
    "    plt.title(f\"Transformed {transformation_name} - Actual vs Predicted DNA Methylation Counts\")\n",
    "    plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data\n",
    "gene_matrix_array = np.load(\"gene_matrix_list.npy\")\n",
    "rna_expression_df = pd.read_csv(\"rna_expression_list.csv\")\n",
    "\n",
    "# Check order of genes in both files is same\n",
    "assert gene_matrix_array.shape[0] == len(\n",
    "    rna_expression_df\n",
    "), \"Mismatch in number of genes\"\n",
    "\n",
    "# Separate mod types\n",
    "h3k9me3_features = gene_matrix_array[:, :, 1]  # Shape (58780, 4000)\n",
    "h3k27me3_features = gene_matrix_array[:, :, 2]\n",
    "dnam_features = gene_matrix_array[:, :, 0]  # Shape (58780, 4000)\n",
    "\n",
    "# Concat histone mods and gene expression to create features\n",
    "X_histone = np.concatenate(\n",
    "    (h3k9me3_features, h3k27me3_features), axis=1\n",
    ")  # Shape (58780, 8000)\n",
    "X_expression = rna_expression_df[\"expression\"].values.reshape(\n",
    "    -1, 1\n",
    ")  # Shape (58780, 1)\n",
    "X = np.concatenate((X_histone, X_expression), axis=1)  # Shape (58780, 8001)\n",
    "\n",
    "# Set DNA methylation as target - altering y to become a count of 1s in each 4000 length array\n",
    "y = np.sum(dnam_features, axis=1)  # Shape (58780,)\n",
    "\n",
    "# Split data into training and testing\n",
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "    X, y, test_size=0.2, random_state=42\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### LGBM regression to predict methylation count (across 4000 bases) on K9, K27 and gene expression counts\n",
    "\n",
    "from lightgbm import LGBMRegressor\n",
    "\n",
    "# Train and eval model\n",
    "lgbm = LGBMRegressor(random_state=42)\n",
    "lgbm.fit(X_train, y_train)\n",
    "y_pred = lgbm.predict(X_test)\n",
    "\n",
    "# Calc and print metrics\n",
    "mae = mean_absolute_error(y_test, y_pred)\n",
    "mse = mean_squared_error(y_test, y_pred)\n",
    "r2 = r2_score(y_test, y_pred)\n",
    "\n",
    "print(\"Random Forest Regressor metrics:\")\n",
    "print(\"Mean Absolute Error (MAE):\", mae)\n",
    "print(\"Mean Squared Error (MSE):\", mse)\n",
    "print(\"R-squared (R2):\", r2)\n",
    "\n",
    "# Plot actual vs predicted counts\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.scatter(y_test, y_pred, alpha=0.5)\n",
    "plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], \"--r\")\n",
    "plt.xlabel(\"Actual Counts\")\n",
    "plt.ylabel(\"Predicted Counts\")\n",
    "plt.title(\"LGBM Actual vs Predicted DNA Methylation Counts\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### LGB ALT PLOTS\n",
    "# Log transformation\n",
    "y_log = np.log1p(y_test)  # log(1 + y) to handle log(0)\n",
    "y_pred_log = np.log1p(y_pred)\n",
    "plot_transformations(y_log, y_pred_log, \"Log\")\n",
    "\n",
    "# Square root transformation\n",
    "y_sqrt = np.sqrt(y_test)\n",
    "y_pred_sqrt = np.sqrt(y_pred)\n",
    "plot_transformations(y_sqrt, y_pred_sqrt, \"Square Root\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### RF regression to predict methylation count (across 4000 bases) on K9, K27 and gene expression counts\n",
    "\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "\n",
    "\n",
    "\n",
    "# Train and eval model\n",
    "rf = RandomForestRegressor(random_state=42)\n",
    "rf.fit(X_train, y_train)\n",
    "y_pred = rf.predict(X_test)\n",
    "\n",
    "# Calc and print metrics\n",
    "mae = mean_absolute_error(y_test, y_pred)\n",
    "mse = mean_squared_error(y_test, y_pred)\n",
    "r2 = r2_score(y_test, y_pred)\n",
    "\n",
    "print(\"LGBM Regressor metrics:\")\n",
    "print(\"Mean Absolute Error (MAE):\", mae)\n",
    "print(\"Mean Squared Error (MSE):\", mse)\n",
    "print(\"R-squared (R2):\", r2)\n",
    "\n",
    "# Plot actual vs predicted counts\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.scatter(y_test, y_pred, alpha=0.5)\n",
    "plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], \"--r\")\n",
    "plt.xlabel(\"Actual Counts\")\n",
    "plt.ylabel(\"Predicted Counts\")\n",
    "plt.title(\"RF Actual vs Predicted DNA Methylation Counts\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### RF ALT PLOTS\n",
    "# Log transformation\n",
    "y_log = np.log1p(y_test)  # log(1 + y) to handle log(0)\n",
    "y_pred_log = np.log1p(y_pred)\n",
    "plot_transformations(y_log, y_pred_log, \"Log\")\n",
    "\n",
    "# Square root transformation\n",
    "y_sqrt = np.sqrt(y_test)\n",
    "y_pred_sqrt = np.sqrt(y_pred)\n",
    "plot_transformations(y_sqrt, y_pred_sqrt, \"Square Root\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### SVM regression to predict methylation count (across 4000 bases) on K9, K27 and gene expression counts\n",
    "\n",
    "from sklearn.svm import SVR\n",
    "\n",
    "\n",
    "# Train and eval model\n",
    "svr = SVR()\n",
    "svr.fit(X_train, y_train)\n",
    "y_pred = svr.predict(X_test)\n",
    "\n",
    "# Calc and print metrics\n",
    "mae = mean_absolute_error(y_test, y_pred)\n",
    "mse = mean_squared_error(y_test, y_pred)\n",
    "r2 = r2_score(y_test, y_pred)\n",
    "\n",
    "print(\"SVM Regressor metrics:\")\n",
    "print(\"Mean Absolute Error (MAE):\", mae)\n",
    "print(\"Mean Squared Error (MSE):\", mse)\n",
    "print(\"R-squared (R2):\", r2)\n",
    "\n",
    "# Plot actual vs predicted counts\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.scatter(y_test, y_pred, alpha=0.5)\n",
    "plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], \"--r\")\n",
    "plt.xlabel(\"Actual Counts\")\n",
    "plt.ylabel(\"Predicted Counts\")\n",
    "plt.title(\"SVM Actual vs Predicted DNA Methylation Counts\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### SVM ALT PLOTS\n",
    "\n",
    "# Log transformation\n",
    "y_log = np.log1p(y_test)  # log(1 + y) to handle log(0)\n",
    "y_pred_log = np.log1p(y_pred)\n",
    "plot_transformations(y_log, y_pred_log, \"Log\")\n",
    "\n",
    "# Square root transformation\n",
    "y_sqrt = np.sqrt(y_test)\n",
    "y_pred_sqrt = np.sqrt(y_pred)\n",
    "plot_transformations(y_sqrt, y_pred_sqrt, \"Square Root\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### XGB regression to predict methylation count (across 4000 bases) on K9, K27 and gene expression counts\n",
    "\n",
    "from xgboost import XGBRegressor\n",
    "\n",
    "# Train and eval model\n",
    "xgb = XGBRegressor(random_state=42)\n",
    "xgb.fit(X_train, y_train)\n",
    "y_pred = xgb.predict(X_test)\n",
    "\n",
    "# Calc and print metrics\n",
    "mae = mean_absolute_error(y_test, y_pred)\n",
    "mse = mean_squared_error(y_test, y_pred)\n",
    "r2 = r2_score(y_test, y_pred)\n",
    "\n",
    "print(\"XGB Regressor metrics:\")\n",
    "print(\"Mean Absolute Error (MAE):\", mae)\n",
    "print(\"Mean Squared Error (MSE):\", mse)\n",
    "print(\"R-squared (R2):\", r2)\n",
    "\n",
    "# Plot actual vs predicted counts\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.scatter(y_test, y_pred, alpha=0.5)\n",
    "plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], \"--r\")\n",
    "plt.xlabel(\"Actual Counts\")\n",
    "plt.ylabel(\"Predicted Counts\")\n",
    "plt.title(\"XGB Actual vs Predicted DNA Methylation Counts\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### XGB ALT PLOTS\n",
    "\n",
    "# Log transformation\n",
    "y_log = np.log1p(y_test)  # log(1 + y) to handle log(0)\n",
    "y_pred_log = np.log1p(y_pred)\n",
    "plot_transformations(y_log, y_pred_log, \"Log\")\n",
    "\n",
    "# Square root transformation\n",
    "y_sqrt = np.sqrt(y_test)\n",
    "y_pred_sqrt = np.sqrt(y_pred)\n",
    "plot_transformations(y_sqrt, y_pred_sqrt, \"Square Root\")"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
